<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop</title>
  <meta name="description" content="Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop" />
  
  
  

<meta name="author" content="Stephanie Hicks (stephaniehicks)">


<meta name="date" content="2018-12-03">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="cleaning-the-expression-matrix.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Welcome!</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#preface"><i class="fa fa-check"></i><b>1.1</b> Preface</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#about-the-course"><i class="fa fa-check"></i><b>1.2</b> About the course</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.3</b> GitHub</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.4</b> Installation</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.6</b> Prerequisites</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.7</b> Contact</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html"><i class="fa fa-check"></i><b>2</b> Introduction to single-cell RNA-seq</a><ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#bulk-rna-seq"><i class="fa fa-check"></i><b>2.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#scrna-seq"><i class="fa fa-check"></i><b>2.2</b> scRNA-seq</a></li>
<li class="chapter" data-level="2.3" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#workflow"><i class="fa fa-check"></i><b>2.3</b> Workflow</a></li>
<li class="chapter" data-level="2.4" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#computational-analysis"><i class="fa fa-check"></i><b>2.4</b> Computational Analysis</a></li>
<li class="chapter" data-level="2.5" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#challenges"><i class="fa fa-check"></i><b>2.5</b> Challenges</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#experimental-methods"><i class="fa fa-check"></i><b>2.6</b> Experimental methods</a></li>
<li class="chapter" data-level="2.7" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#what-platform-to-use-for-my-experiment"><i class="fa fa-check"></i><b>2.7</b> What platform to use for my experiment?</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html"><i class="fa fa-check"></i><b>3</b> Construction of expression matrix</a><ul>
<li class="chapter" data-level="3.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#raw-sequencing-reads-qc"><i class="fa fa-check"></i><b>3.1</b> Raw sequencing reads QC</a><ul>
<li class="chapter" data-level="3.1.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#fastq"><i class="fa fa-check"></i><b>3.1.1</b> FastQ</a></li>
<li class="chapter" data-level="3.1.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#check-the-quality-of-the-raw-sequencing-reads"><i class="fa fa-check"></i><b>3.1.2</b> Check the quality of the raw sequencing reads</a></li>
<li class="chapter" data-level="3.1.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#trim-sequencing-adapaters-from-sequencing-reads"><i class="fa fa-check"></i><b>3.1.3</b> Trim sequencing adapaters from sequencing reads</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-alignment"><i class="fa fa-check"></i><b>3.2</b> Reads alignment</a><ul>
<li class="chapter" data-level="3.2.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#genome-fasta-gtf"><i class="fa fa-check"></i><b>3.2.1</b> Genome (<code>FASTA</code>, <code>GTF</code>)</a></li>
<li class="chapter" data-level="3.2.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#alignment-using-star"><i class="fa fa-check"></i><b>3.2.2</b> Alignment using Star</a></li>
<li class="chapter" data-level="3.2.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#pseudo-alignment-using-salmon"><i class="fa fa-check"></i><b>3.2.3</b> Pseudo-alignment using Salmon</a></li>
<li class="chapter" data-level="3.2.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapped-reads-bam-file-format"><i class="fa fa-check"></i><b>3.2.4</b> Mapped reads: <code>BAM</code> file format</a></li>
<li class="chapter" data-level="3.2.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cram"><i class="fa fa-check"></i><b>3.2.5</b> CRAM</a></li>
<li class="chapter" data-level="3.2.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-mapping-qc"><i class="fa fa-check"></i><b>3.2.6</b> Reads Mapping QC</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-quantification"><i class="fa fa-check"></i><b>3.3</b> Reads quantification</a></li>
<li class="chapter" data-level="3.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#unique-molecular-identifiers-umis"><i class="fa fa-check"></i><b>3.4</b> Unique Molecular Identifiers (UMIs)</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html"><i class="fa fa-check"></i><b>4</b> Introduction to R/Bioconductor</a><ul>
<li class="chapter" data-level="4.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#installing-r-packages"><i class="fa fa-check"></i><b>4.1</b> Installing R packages</a><ul>
<li class="chapter" data-level="4.1.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#cran"><i class="fa fa-check"></i><b>4.1.1</b> CRAN</a></li>
<li class="chapter" data-level="4.1.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#github-1"><i class="fa fa-check"></i><b>4.1.2</b> Github</a></li>
<li class="chapter" data-level="4.1.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#bioconductor"><i class="fa fa-check"></i><b>4.1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="4.1.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#source"><i class="fa fa-check"></i><b>4.1.4</b> Source</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#data-types"><i class="fa fa-check"></i><b>4.2</b> Data Types</a><ul>
<li class="chapter" data-level="4.2.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#what-is-tidy-data"><i class="fa fa-check"></i><b>4.2.1</b> What is Tidy Data?</a></li>
<li class="chapter" data-level="4.2.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#what-is-rich-data"><i class="fa fa-check"></i><b>4.2.2</b> What is Rich Data?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#bioconductor-1"><i class="fa fa-check"></i><b>4.3</b> Bioconductor</a><ul>
<li class="chapter" data-level="4.3.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#singlecellexperiment-class"><i class="fa fa-check"></i><b>4.3.1</b> <code>SingleCellExperiment</code> class</a></li>
<li class="chapter" data-level="4.3.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#scater-package"><i class="fa fa-check"></i><b>4.3.2</b> <code>scater</code> package</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#data-visualization-using-ggplot2"><i class="fa fa-check"></i><b>4.4</b> Data visualization using ggplot2</a><ul>
<li class="chapter" data-level="4.4.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#what-is-ggplot2"><i class="fa fa-check"></i><b>4.4.1</b> What is ggplot2?</a></li>
<li class="chapter" data-level="4.4.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#principles-of-ggplot2"><i class="fa fa-check"></i><b>4.4.2</b> Principles of ggplot2</a></li>
<li class="chapter" data-level="4.4.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#using-the-aes-mapping-function"><i class="fa fa-check"></i><b>4.4.3</b> Using the <code>aes()</code> mapping function</a></li>
<li class="chapter" data-level="4.4.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#geoms"><i class="fa fa-check"></i><b>4.4.4</b> Geoms</a></li>
<li class="chapter" data-level="4.4.5" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#plotting-data-from-more-than-2-cells"><i class="fa fa-check"></i><b>4.4.5</b> Plotting data from more than 2 cells</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html"><i class="fa fa-check"></i><b>5</b> Cleaning the expression matrix</a><ul>
<li class="chapter" data-level="5.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#expression-qc"><i class="fa fa-check"></i><b>5.1</b> Expression QC</a><ul>
<li class="chapter" data-level="5.1.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction"><i class="fa fa-check"></i><b>5.1.1</b> Introduction</a></li>
<li class="chapter" data-level="5.1.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#tung-dataset-umi"><i class="fa fa-check"></i><b>5.1.2</b> Tung dataset (UMI)</a></li>
<li class="chapter" data-level="5.1.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cell-qc"><i class="fa fa-check"></i><b>5.1.3</b> Cell QC</a></li>
<li class="chapter" data-level="5.1.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cell-filtering"><i class="fa fa-check"></i><b>5.1.4</b> Cell filtering</a></li>
<li class="chapter" data-level="5.1.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#compare-filterings"><i class="fa fa-check"></i><b>5.1.5</b> Compare filterings</a></li>
<li class="chapter" data-level="5.1.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#gene-analysis"><i class="fa fa-check"></i><b>5.1.6</b> Gene analysis</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#data-visualization"><i class="fa fa-check"></i><b>5.2</b> Data visualization</a><ul>
<li class="chapter" data-level="5.2.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-1"><i class="fa fa-check"></i><b>5.2.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#pca-plot"><i class="fa fa-check"></i><b>5.2.2</b> PCA plot</a></li>
<li class="chapter" data-level="5.2.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#tsne-map"><i class="fa fa-check"></i><b>5.2.3</b> tSNE map</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#identifying-confounding-factors"><i class="fa fa-check"></i><b>5.3</b> Identifying confounding factors</a><ul>
<li class="chapter" data-level="5.3.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-2"><i class="fa fa-check"></i><b>5.3.1</b> Introduction</a></li>
<li class="chapter" data-level="5.3.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#correlations-with-pcs"><i class="fa fa-check"></i><b>5.3.2</b> Correlations with PCs</a></li>
<li class="chapter" data-level="5.3.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#explanatory-variables"><i class="fa fa-check"></i><b>5.3.3</b> Explanatory variables</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalization-theory"><i class="fa fa-check"></i><b>5.4</b> Normalization theory</a><ul>
<li class="chapter" data-level="5.4.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-3"><i class="fa fa-check"></i><b>5.4.1</b> Introduction</a></li>
<li class="chapter" data-level="5.4.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#library-size-1"><i class="fa fa-check"></i><b>5.4.2</b> Library size</a></li>
<li class="chapter" data-level="5.4.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalisations"><i class="fa fa-check"></i><b>5.4.3</b> Normalisations</a></li>
<li class="chapter" data-level="5.4.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#effectiveness"><i class="fa fa-check"></i><b>5.4.4</b> Effectiveness</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalization-practice"><i class="fa fa-check"></i><b>5.5</b> Normalization practice</a><ul>
<li class="chapter" data-level="5.5.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#raw"><i class="fa fa-check"></i><b>5.5.1</b> Raw</a></li>
<li class="chapter" data-level="5.5.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cpm-1"><i class="fa fa-check"></i><b>5.5.2</b> CPM</a></li>
<li class="chapter" data-level="5.5.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#scran-1"><i class="fa fa-check"></i><b>5.5.3</b> scran</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#dealing-with-confounders"><i class="fa fa-check"></i><b>5.6</b> Dealing with confounders</a><ul>
<li class="chapter" data-level="5.6.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-4"><i class="fa fa-check"></i><b>5.6.1</b> Introduction</a></li>
<li class="chapter" data-level="5.6.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#combat"><i class="fa fa-check"></i><b>5.6.2</b> Combat</a></li>
<li class="chapter" data-level="5.6.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#mnncorrect"><i class="fa fa-check"></i><b>5.6.3</b> mnnCorrect</a></li>
<li class="chapter" data-level="5.6.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#glm"><i class="fa fa-check"></i><b>5.6.4</b> GLM</a></li>
<li class="chapter" data-level="5.6.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#how-to-evaluate-and-compare-confounder-removal-strategies"><i class="fa fa-check"></i><b>5.6.5</b> How to evaluate and compare confounder removal strategies</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="biological-analysis.html"><a href="biological-analysis.html"><i class="fa fa-check"></i><b>6</b> Biological Analysis</a><ul>
<li class="chapter" data-level="6.1" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering"><i class="fa fa-check"></i><b>6.1</b> Clustering</a><ul>
<li class="chapter" data-level="6.1.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-5"><i class="fa fa-check"></i><b>6.1.1</b> Introduction</a></li>
<li class="chapter" data-level="6.1.2" data-path="biological-analysis.html"><a href="biological-analysis.html#dimensionality-reduction"><i class="fa fa-check"></i><b>6.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.1.3" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-methods"><i class="fa fa-check"></i><b>6.1.3</b> Clustering methods</a></li>
<li class="chapter" data-level="6.1.4" data-path="biological-analysis.html"><a href="biological-analysis.html#challenges-in-clustering"><i class="fa fa-check"></i><b>6.1.4</b> Challenges in clustering</a></li>
<li class="chapter" data-level="6.1.5" data-path="biological-analysis.html"><a href="biological-analysis.html#software-packages-for-scrna-seq-data"><i class="fa fa-check"></i><b>6.1.5</b> Software packages for scRNA-seq data</a></li>
<li class="chapter" data-level="6.1.6" data-path="biological-analysis.html"><a href="biological-analysis.html#comparing-clustering"><i class="fa fa-check"></i><b>6.1.6</b> Comparing clustering</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-example"><i class="fa fa-check"></i><b>6.2</b> Clustering example</a><ul>
<li class="chapter" data-level="6.2.1" data-path="biological-analysis.html"><a href="biological-analysis.html#deng-dataset"><i class="fa fa-check"></i><b>6.2.1</b> Deng dataset</a></li>
<li class="chapter" data-level="6.2.2" data-path="biological-analysis.html"><a href="biological-analysis.html#sc3-1"><i class="fa fa-check"></i><b>6.2.2</b> SC3</a></li>
<li class="chapter" data-level="6.2.3" data-path="biological-analysis.html"><a href="biological-analysis.html#tsne-kmeans"><i class="fa fa-check"></i><b>6.2.3</b> tSNE + kmeans</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="biological-analysis.html"><a href="biological-analysis.html#feature-selection"><i class="fa fa-check"></i><b>6.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="6.3.1" data-path="biological-analysis.html"><a href="biological-analysis.html#identifying-genes-vs-a-null-model"><i class="fa fa-check"></i><b>6.3.1</b> Identifying Genes vs a Null Model</a></li>
<li class="chapter" data-level="6.3.2" data-path="biological-analysis.html"><a href="biological-analysis.html#correlated-expression"><i class="fa fa-check"></i><b>6.3.2</b> Correlated Expression</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="biological-analysis.html"><a href="biological-analysis.html#differential-expression-de-analysis"><i class="fa fa-check"></i><b>6.4</b> Differential Expression (DE) analysis</a><ul>
<li class="chapter" data-level="6.4.1" data-path="biological-analysis.html"><a href="biological-analysis.html#bulk-rna-seq-1"><i class="fa fa-check"></i><b>6.4.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="6.4.2" data-path="biological-analysis.html"><a href="biological-analysis.html#single-cell-rna-seq"><i class="fa fa-check"></i><b>6.4.2</b> Single-cell RNA-seq</a></li>
<li class="chapter" data-level="6.4.3" data-path="biological-analysis.html"><a href="biological-analysis.html#differences-in-distribution"><i class="fa fa-check"></i><b>6.4.3</b> Differences in distribution</a></li>
<li class="chapter" data-level="6.4.4" data-path="biological-analysis.html"><a href="biological-analysis.html#models-of-single-cell-rna-seq-data"><i class="fa fa-check"></i><b>6.4.4</b> Models of single-cell RNA-seq data</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="biological-analysis.html"><a href="biological-analysis.html#de-in-tung-dataset"><i class="fa fa-check"></i><b>6.5</b> DE in <code>tung</code> dataset</a><ul>
<li class="chapter" data-level="6.5.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-6"><i class="fa fa-check"></i><b>6.5.1</b> Introduction</a></li>
<li class="chapter" data-level="6.5.2" data-path="biological-analysis.html"><a href="biological-analysis.html#kolmogorov-smirnov-test"><i class="fa fa-check"></i><b>6.5.2</b> Kolmogorov-Smirnov test</a></li>
<li class="chapter" data-level="6.5.3" data-path="biological-analysis.html"><a href="biological-analysis.html#wilcoxmann-whitney-u-test"><i class="fa fa-check"></i><b>6.5.3</b> Wilcox/Mann-Whitney-U Test</a></li>
<li class="chapter" data-level="6.5.4" data-path="biological-analysis.html"><a href="biological-analysis.html#edger"><i class="fa fa-check"></i><b>6.5.4</b> edgeR</a></li>
<li class="chapter" data-level="6.5.5" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle"><i class="fa fa-check"></i><b>6.5.5</b> Monocle</a></li>
<li class="chapter" data-level="6.5.6" data-path="biological-analysis.html"><a href="biological-analysis.html#mast"><i class="fa fa-check"></i><b>6.5.6</b> MAST</a></li>
<li class="chapter" data-level="6.5.7" data-path="biological-analysis.html"><a href="biological-analysis.html#scde"><i class="fa fa-check"></i><b>6.5.7</b> SCDE</a></li>
<li class="chapter" data-level="6.5.8" data-path="biological-analysis.html"><a href="biological-analysis.html#zinbwave-deseq2"><i class="fa fa-check"></i><b>6.5.8</b> zinbwave + DESeq2</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="http://www.sanger.ac.uk/science/groups/hemberg-group" target="blank">Hemberg Lab</a></li>
<li><a href="http://www.stephaniehicks.comp" target="blank">Stephanie Hicks</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="biological-analysis" class="section level1">
<h1><span class="header-section-number">6</span> Biological Analysis</h1>
<div id="clustering" class="section level2">
<h2><span class="header-section-number">6.1</span> Clustering</h2>
<p>Once we have normalized the data and removed confounders we can
carry out analyses that are relevant to the biological questions
at hand. The exact nature of the analysis depends on the dataset.
Nevertheless, there are a few aspects that are useful in a wide
range of contexts and we will be discussing some of them in the
next few chapters. We will start with the clustering of
scRNA-seq data.</p>
<div id="introduction-5" class="section level3">
<h3><span class="header-section-number">6.1.1</span> Introduction</h3>
<p>One of the most promising applications of scRNA-seq is <em>de novo</em>
discovery and annotation of cell-types based on transcription
profiles. Computationally, this is a hard problem as it amounts to
<strong>unsupervised clustering</strong>. That is, we need to identify groups of
cells based on the similarities of the transcriptomes without any
prior knowledge of the labels. Moreover, in most situations we
do not even know the number of clusters <em>a priori</em>. The problem
is made even more challenging due to the high level of noise
(both technical and biological) and the large number of dimensions
(i.e. genes).</p>
</div>
<div id="dimensionality-reduction" class="section level3">
<h3><span class="header-section-number">6.1.2</span> Dimensionality reduction</h3>
<p>When working with large datasets, it can often be beneficial to
apply some sort of dimensionality reduction method. By projecting
the data onto a lower-dimensional sub-space, one is often able to
significantly reduce the amount of noise. An additional benefit is
that it is typically much easier to visualize the data in a 2 or
3-dimensional subspace. We have already discussed PCA
and t-SNE.</p>
</div>
<div id="clustering-methods" class="section level3">
<h3><span class="header-section-number">6.1.3</span> Clustering methods</h3>
<p><strong>Unsupervised clustering</strong> is useful in many different applications and
it has been widely studied in machine learning. Some of the most
popular approaches are <strong>hierarchical clustering</strong>,
<strong>k-means clustering</strong> and <strong>graph-based clustering</strong>.</p>
<div id="hierarchical-clustering" class="section level4">
<h4><span class="header-section-number">6.1.3.1</span> Hierarchical clustering</h4>
<p>In <a href="https://en.wikipedia.org/wiki/Hierarchical_clustering">hierarchical clustering</a>,
one can use either a bottom-up or a top-down approach. In the former
case, each cell is initially assigned to its own cluster and
pairs of clusters are subsequently merged to create a hieararchy:</p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb132-1" title="1">knitr<span class="op">::</span><span class="kw">include_graphics</span>(<span class="st">&quot;figures/hierarchical_clustering2.png&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:clust-hierarch-dendr"></span>
<img src="figures/hierarchical_clustering2.png" alt="The hierarchical clustering dendrogram" width="50%" />
<p class="caption">
Figure 6.1: The hierarchical clustering dendrogram
</p>
</div>
<p>With a top-down strategy, one instead starts with all
observations in one cluster and then recursively split each
cluster to form a hierarchy. One of the advantages of
this strategy is that the method is deterministic.</p>
</div>
<div id="k-means" class="section level4">
<h4><span class="header-section-number">6.1.3.2</span> k-means</h4>
<p>In <a href="https://en.wikipedia.org/wiki/K-means_clustering"><em>k</em>-means clustering</a>,
the goal is to partition <em>N</em> cells into <em>k</em> different clusters.
In an iterative manner, cluster centers are assigned and
each cell is assigned to its nearest cluster:</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb133-1" title="1">knitr<span class="op">::</span><span class="kw">include_graphics</span>(<span class="st">&quot;figures/k-means.png&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:clust-k-means"></span>
<img src="figures/k-means.png" alt="Schematic representation of the k-means clustering" width="100%" />
<p class="caption">
Figure 6.2: Schematic representation of the k-means clustering
</p>
</div>
<p>Most methods for scRNA-seq analysis includes a <em>k</em>-means
step at some point.</p>
</div>
<div id="graph-based-methods" class="section level4">
<h4><span class="header-section-number">6.1.3.3</span> Graph-based methods</h4>
<p>Over the last two decades there has been a lot of interest in
analyzing networks in various domains. One goal is to
identify groups or modules of nodes in a network.</p>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb134-1" title="1">knitr<span class="op">::</span><span class="kw">include_graphics</span>(<span class="st">&quot;figures/graph_network.jpg&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:clust-graph"></span>
<img src="figures/graph_network.jpg" alt="Schematic representation of the graph network" width="100%" />
<p class="caption">
Figure 6.3: Schematic representation of the graph network
</p>
</div>
<p>Some of these methods can be applied to scRNA-seq data
by building a graph where each node represents a cell.
Note that constructing the graph and assigning weights
to the edges is not trivial. One advantage of
graph-based methods is that some of them are very
efficient and can be applied to networks containing
millions of nodes.</p>
</div>
</div>
<div id="challenges-in-clustering" class="section level3">
<h3><span class="header-section-number">6.1.4</span> Challenges in clustering</h3>
<ul>
<li>What is the number of clusters <em>k</em>?</li>
<li>What is a cell type?</li>
<li><strong>Scalability</strong>: in the last few years the number of cells in scRNA-seq experiments has grown by several orders of magnitude from ~<span class="math inline">\(10^2\)</span> to ~<span class="math inline">\(10^6\)</span></li>
<li>Tools are not user-friendly</li>
</ul>
</div>
<div id="software-packages-for-scrna-seq-data" class="section level3">
<h3><span class="header-section-number">6.1.5</span> Software packages for scRNA-seq data</h3>
<div id="clusterexperiment" class="section level4">
<h4><span class="header-section-number">6.1.5.1</span> <a href="https://www.bioconductor.org/packages/release/bioc/html/clusterExperiment.html">clusterExperiment</a></h4>
<ul>
<li>R/Bioconductor package uses <code>SingleCellExperiment</code> object</li>
<li>Functionality for running and comparing many different clusterings of single-cell sequencing data or other large mRNA expression data sets</li>
<li>Based on the RSEC (Resampling-based Sequential Ensemble Clustering) algorithm for finding a single robust clustering based on the many clusterings that the user might create by perturbing various parameters of a clustering algorithm</li>
</ul>
</div>
<div id="sc3" class="section level4">
<h4><span class="header-section-number">6.1.5.2</span> <a href="http://bioconductor.org/packages/SC3/">SC3</a></h4>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb135-1" title="1">knitr<span class="op">::</span><span class="kw">include_graphics</span>(<span class="st">&quot;figures/sc3.png&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:clust-sc3"></span>
<img src="figures/sc3.png" alt="SC3 pipeline" width="100%" />
<p class="caption">
Figure 6.4: SC3 pipeline
</p>
</div>
<ul>
<li>R/Bioconductor package based on PCA and spectral dimensionality reductions <span class="citation">(Kiselev et al. <a href="#ref-Kiselev2016-bq">2017</a>)</span></li>
<li>Utilises <em>k</em>-means</li>
<li>Additionally performs the consensus clustering</li>
</ul>
</div>
<div id="tsne-k-means" class="section level4">
<h4><span class="header-section-number">6.1.5.3</span> tSNE + k-means</h4>
<ul>
<li>Based on <strong>tSNE</strong> maps</li>
<li>Utilises <em>k</em>-means</li>
</ul>
</div>
<div id="sincera" class="section level4">
<h4><span class="header-section-number">6.1.5.4</span> <a href="https://research.cchmc.org/pbge/sincera.html">SINCERA</a></h4>
<ul>
<li>SINCERA <span class="citation">(Guo et al. <a href="#ref-Guo2015-ok">2015</a>)</span> is based on hierarchical clustering</li>
<li>Data is converted to <em>z</em>-scores before clustering</li>
<li>Identify <em>k</em> by finding the first singleton cluster in the hierarchy</li>
</ul>
</div>
<div id="pcareduce" class="section level4">
<h4><span class="header-section-number">6.1.5.5</span> <a href="https://github.com/JustinaZ/pcaReduce">pcaReduce</a></h4>
<p>pcaReduce <span class="citation">(žurauskienė and Yau <a href="#ref-Zurauskiene2016-kg">2016</a>)</span> combines PCA, <em>k</em>-means and “iterative” hierarchical clustering. Starting from a large number of clusters pcaReduce iteratively merges similar clusters; after each merging event it removes the principle component explaning the least variance in the data.</p>
</div>
<div id="snn-cliq" class="section level4">
<h4><span class="header-section-number">6.1.5.6</span> <a href="http://bioinfo.uncc.edu/SNNCliq/">SNN-Cliq</a></h4>
<p><code>SNN-Cliq</code> <span class="citation">(Xu and Su <a href="#ref-Xu2015-vf">2015</a>)</span> is a graph-based method. First the method identifies the k-nearest-neighbours of each cell according to the <em>distance</em> measure. This is used to calculate the number of Shared Nearest Neighbours (SNN) between each pair of cells. A graph is built by placing an edge between two cells If they have at least one SNN. Clusters are defined as groups of cells with many edges between them using a “clique” method. SNN-Cliq requires several parameters to be defined manually.</p>
</div>
<div id="seurat-clustering" class="section level4">
<h4><span class="header-section-number">6.1.5.7</span> Seurat clustering</h4>
<p><a href="https://github.com/satijalab/seurat"><code>Seurat</code></a> clustering is based on a <em>community detection</em> approach similar to <code>SNN-Cliq</code> and to one previously proposed for analyzing CyTOF data <span class="citation">(Levine et al. <a href="#ref-Levine2015-fk">2015</a>)</span>. Since <code>Seurat</code> has become more like an all-in-one tool for scRNA-seq data analysis we dedicate a separate chapter to discuss it in more details (chapter <a href="#seurat-chapter"><strong>??</strong></a>).</p>
</div>
</div>
<div id="comparing-clustering" class="section level3">
<h3><span class="header-section-number">6.1.6</span> Comparing clustering</h3>
<p>To compare two sets of clustering labels we can use <a href="https://en.wikipedia.org/wiki/Rand_index">adjusted Rand index</a>. The index is a measure of the similarity between two data clusterings. Values of the adjusted Rand index lie in <span class="math inline">\([0;1]\)</span> interval, where <span class="math inline">\(1\)</span> means that two clusterings are identical and <span class="math inline">\(0\)</span> means the level of similarity expected by chance.</p>
</div>
</div>
<div id="clustering-example" class="section level2">
<h2><span class="header-section-number">6.2</span> Clustering example</h2>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb136-1" title="1"><span class="kw">library</span>(SC3)</a></code></pre></div>
<p>To illustrate clustering of scRNA-seq data, we consider
the <code>Deng</code> dataset of cells from developing mouse embryo
<span class="citation">(Deng et al. <a href="#ref-Deng2014-mx">2014</a>)</span>. We have preprocessed the dataset and
created a <code>SingleCellExperiment</code> object in advance.
We have also annotated the cells with the cell types
identified in the original publication (it is the
<code>cell_type2</code> column in the <code>colData</code> slot).</p>
<div id="deng-dataset" class="section level3">
<h3><span class="header-section-number">6.2.1</span> Deng dataset</h3>
<p>Let’s load the data and look at it:</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb137-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/deng/deng-reads.rds&quot;</span>)</a>
<a class="sourceLine" id="cb137-2" title="2">deng</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 22431 268 
## metadata(0):
## assays(2): counts logcounts
## rownames(22431): Hvcn1 Gbp7 ... Sox5 Alg11
## rowData names(10): feature_symbol is_feature_control ...
##   total_counts log10_total_counts
## colnames(268): 16cell 16cell.1 ... zy.2 zy.3
## colData names(30): cell_type2 cell_type1 ... pct_counts_ERCC
##   is_cell_control
## reducedDimNames(0):
## spikeNames(1): ERCC</code></pre>
<p>Let’s look at the cell type annotation:</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb139-1" title="1"><span class="kw">table</span>(<span class="kw">colData</span>(deng)<span class="op">$</span>cell_type2)</a></code></pre></div>
<pre><code>## 
##     16cell      4cell      8cell early2cell earlyblast  late2cell 
##         50         14         37          8         43         10 
##  lateblast   mid2cell   midblast         zy 
##         30         12         60          4</code></pre>
<p>A simple PCA analysis already separates some strong
cell types and provides some insights in the data structure:</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb141-1" title="1"><span class="kw">plotPCA</span>(deng, <span class="dt">colour_by =</span> <span class="st">&quot;cell_type2&quot;</span>)</a></code></pre></div>
<p><img src="06-biological-analyses_files/figure-html/unnamed-chunk-5-1.png" width="90%" style="display: block; margin: auto;" /></p>
<p>As you can see, the early cell types separate quite well,
but the three blastocyst timepoints are more difficult to distinguish.</p>
</div>
<div id="sc3-1" class="section level3">
<h3><span class="header-section-number">6.2.2</span> SC3</h3>
<p>Let’s run <code>SC3</code> clustering on the Deng data. The advantage
of the <code>SC3</code> is that it can directly ingest a
<code>SingleCellExperiment</code> object.</p>
<p>Now let’s image we do not know the number of clusters <em>k</em>
(cell types). <code>SC3</code> can estimate a number of clusters for you:</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb142-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">sc3_estimate_k</span>(deng)</a></code></pre></div>
<pre><code>## Estimating k...</code></pre>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb144-1" title="1"><span class="kw">metadata</span>(deng)<span class="op">$</span>sc3<span class="op">$</span>k_estimation</a></code></pre></div>
<pre><code>## [1] 6</code></pre>
<p>Interestingly, the number of cell types predicted by <code>SC3</code>
is smaller than in the original data annotation. However,
early, mid and late stages of different cell types
together, we will have exactly 6 cell types. We store
the merged cell types in <code>cell_type1</code> column of the
<code>colData</code> slot:</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb146-1" title="1"><span class="kw">plotPCA</span>(deng, <span class="dt">colour_by =</span> <span class="st">&quot;cell_type1&quot;</span>)</a></code></pre></div>
<p><img src="06-biological-analyses_files/figure-html/unnamed-chunk-7-1.png" width="90%" style="display: block; margin: auto;" /></p>
<p>Now we are ready to run <code>SC3</code> (we also ask it to calculate
biological properties of the clusters)</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb147-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">sc3</span>(deng, <span class="dt">ks =</span> <span class="dv">10</span>, <span class="dt">biology =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><code>SC3</code> result consists of several different outputs
(please look in <span class="citation">(Kiselev et al. <a href="#ref-Kiselev2016-bq">2017</a>)</span> and
<a href="http://bioconductor.org/packages/release/bioc/vignettes/SC3/inst/doc/my-vignette.html">SC3 vignette</a>
for more details). Here we show some of them:</p>
<p>First up we plot a “consensus matrix”. The consensus matrix is a
<code>NxN</code> matrix, where <code>N</code> is the number of cells. It represents similarity
between the cells based on the averaging of clustering results from all
combinations of clustering parameters.</p>
<p>Similarity 0 (blue) means that the two cells are always assigned to
different clusters. In contrast, similarity 1 (red) means that the two
cells are always assigned to the same cluster. The consensus matrix is
clustered by hierarchical clustering and has a diagonal-block structure.
Intuitively, the perfect clustering is achieved when all diagonal blocks
are completely red and all off-diagonal elements are completely blue.</p>
<p>We can plot the consensus matrix as a heatmap using the <code>sc3_plot_consensus()</code>
in the <code>SC3</code> package.</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb148-1" title="1"><span class="kw">sc3_plot_consensus</span>(deng, <span class="dt">k =</span> <span class="dv">10</span>, <span class="dt">show_pdata =</span> <span class="st">&quot;cell_type2&quot;</span>)</a></code></pre></div>
<p><img src="06-biological-analyses_files/figure-html/unnamed-chunk-9-1.png" width="90%" style="display: block; margin: auto;" /></p>
<p>Next, we can create a heatmap of the gene expression matrix using
the <code>sc3_plot_expression()</code> function in <code>SC3</code>.</p>
<p>The expression heatmap represents the original input expression matrix
(cells in columns and genes in rows) after applying a gene filter. Genes
are clustered by kmeans with <code>k = 100</code> (dendrogram on the left) and the
heatmap represents the expression levels of the gene cluster centers
after log2-scaling. We also ask for <code>k = 10</code> clusters of cells.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb149-1" title="1"><span class="kw">sc3_plot_expression</span>(deng, <span class="dt">k =</span> <span class="dv">10</span>, </a>
<a class="sourceLine" id="cb149-2" title="2">                    <span class="dt">show_pdata =</span> <span class="st">&quot;cell_type2&quot;</span>)</a></code></pre></div>
<p><img src="06-biological-analyses_files/figure-html/unnamed-chunk-10-1.png" width="90%" style="display: block; margin: auto;" /></p>
<p>We can also create PCA plot with clusters identified from <code>SC3</code>:</p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb150-1" title="1"><span class="kw">plotPCA</span>(deng, <span class="dt">colour_by =</span> <span class="st">&quot;sc3_10_clusters&quot;</span>)</a></code></pre></div>
<p><img src="06-biological-analyses_files/figure-html/unnamed-chunk-11-1.png" width="90%" style="display: block; margin: auto;" /></p>
<p>We can compare the results of <code>SC3</code> clustering with the
original publication cell type labels using the
<a href="https://en.wikipedia.org/wiki/Rand_index#Adjusted_Rand_index">Adjusted Rand Index</a>.</p>
<p>The Rand index (named after William M. Rand) is a measure of the similarity
between two data clusterings. The adjusted Rand index is the Rand index
that is adjusted for the chance grouping of elements. Such a correction for chance
establishes a baseline by using the expected similarity of all pair-wise comparisons
between clusterings specified by a random model. The adjusted Rand index is thus
ensured to have a value close to 0.0 for random labeling independently of the
number of clusters and samples and exactly 1.0 when the clusterings are identical.</p>
<p>Here, we use the function <code>adjustedRandIndex()</code>, which is part of the <code>mclust</code>
package:</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb151-1" title="1">mclust<span class="op">::</span><span class="kw">adjustedRandIndex</span>(<span class="kw">colData</span>(deng)<span class="op">$</span>cell_type2,</a>
<a class="sourceLine" id="cb151-2" title="2">                          <span class="kw">colData</span>(deng)<span class="op">$</span>sc3_<span class="dv">10</span>_clusters)</a></code></pre></div>
<pre><code>## [1] 0.6638246</code></pre>
<p><strong>Note</strong> <code>SC3</code> can also be run in an interactive <code>Shiny</code>
session:</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb153-1" title="1"><span class="kw">sc3_interactive</span>(deng)</a></code></pre></div>
<p>This command will open <code>SC3</code> in a web browser.</p>
<p>Before we leave this section, the package vignette points out that
because of direct calculations of distances, the <code>SC3</code> functions will
become very slow when the number of cells is <span class="math inline">\(&gt;5000\)</span>.</p>
</div>
<div id="tsne-kmeans" class="section level3">
<h3><span class="header-section-number">6.2.3</span> tSNE + kmeans</h3>
<p><a href="https://lvdmaaten.github.io/tsne/">tSNE</a> plots that we saw before
when used the <strong>scater</strong> package are made by using the
<a href="https://cran.r-project.org/web/packages/Rtsne/index.html">Rtsne</a> and
<a href="https://cran.r-project.org/web/packages/ggplot2/index.html">ggplot2</a> packages.
Here we will do the same:</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb154-1" title="1">deng &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(deng)</a>
<a class="sourceLine" id="cb154-2" title="2"><span class="kw">plotTSNE</span>(deng)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:clust-tsne"></span>
<img src="06-biological-analyses_files/figure-html/clust-tsne-1.png" alt="tSNE map of the patient data" width="90%" />
<p class="caption">
Figure 6.5: tSNE map of the patient data
</p>
</div>
<p>Note that all points on the plot above are black. This is different
from what we saw before, when the cells were coloured based on the
annotation.</p>
<p><strong>Here we do not have any annotation and all cells come from the same batch, therefore all dots are black.</strong></p>
<p>Now we are going to apply <em>k</em>-means clustering algorithm to the cloud
of points on the tSNE map. How many groups do you see in the cloud?</p>
<p>We will start with <span class="math inline">\(k=8\)</span>:</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb155-1" title="1"><span class="kw">colData</span>(deng)<span class="op">$</span>tSNE_kmeans &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">kmeans</span>(<span class="kw">reducedDim</span>(deng, <span class="st">&quot;TSNE&quot;</span>), <span class="dt">centers =</span> <span class="dv">8</span>)<span class="op">$</span>clust)</a>
<a class="sourceLine" id="cb155-2" title="2">deng &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(deng)</a>
<a class="sourceLine" id="cb155-3" title="3"><span class="kw">plotTSNE</span>(deng, <span class="dt">colour_by =</span> <span class="st">&quot;tSNE_kmeans&quot;</span>)</a></code></pre></div>
<pre><code>## Warning: &#39;add_ticks&#39; is deprecated.
## Use &#39;+ geom_rug(...)&#39; instead.</code></pre>
<div class="figure" style="text-align: center"><span id="fig:clust-tsne-kmeans2"></span>
<img src="06-biological-analyses_files/figure-html/clust-tsne-kmeans2-1.png" alt="tSNE map of the patient data with 8 colored clusters, identified by the k-means clustering algorithm" width="90%" />
<p class="caption">
Figure 6.6: tSNE map of the patient data with 8 colored clusters, identified by the k-means clustering algorithm
</p>
</div>
<p>As you may have noticed, <code>tSNE+kmeans</code> is stochastic
and give different results every time it’s are run. To get a better
overview of the solutions, we need to run the methods multiple times.
<code>SC3</code> is also stochastic, but thanks to the consensus step, it is
more robust and less likely to produce different outcomes.</p>
</div>
</div>
<div id="feature-selection" class="section level2">
<h2><span class="header-section-number">6.3</span> Feature Selection</h2>
<p>scRNA-seq is capable of measuring the expression of many
thousands of genes in every cell. However, in most situations only a
portion of those will show a response to the biological condition of
interest, e.g. differences in cell-type, drivers of differentiation,
respond to an environmental stimulus. Most genes detected in a scRNA-seq
experiment will only be detected at different levels due to technical
noise. One consequence of this is that technical noise and batch
effects can obscure the biological signal of interest.</p>
<p>Thus, it is often advantageous to perform feature selection to remove
those genes which only exhibit technical noise from downstream analysis.
Not only does this generally increase the signal:noise ratio in the data;
it also reduces the computational complexity of analyses, by reducing
the total amount of data to be processed.</p>
<p>For scRNA-seq data, we typically focus on unsupervised methods of feature
selection, which don’t require any a priori information, such as cell-type
labels or biological group, since they are not available, or may be unreliable,
for many experiments. In contrast, differential expression
can be considered a form of supervised feature selection since it uses the
known biological label of each sample to identify features (i.e. genes) which
are expressed at different levels across groups.</p>
<p>In this section, we will briefly cover the concepts of feature selection, but
due to a limit in time, I refer you to the main coures page for more details
on feature selection:</p>
<p><a href="http://hemberg-lab.github.io/scRNA.seq.course/biological-analysis.html#feature-selection" class="uri">http://hemberg-lab.github.io/scRNA.seq.course/biological-analysis.html#feature-selection</a></p>
<div id="identifying-genes-vs-a-null-model" class="section level3">
<h3><span class="header-section-number">6.3.1</span> Identifying Genes vs a Null Model</h3>
<p>There are two main approaches to unsupervised feature selection. The
first is to identify genes which behave differently from a null model
describing just the technical noise expected in the dataset.</p>
<p>If the dataset contains spike-in RNAs they can be used to directly model
technical noise. However, measurements of spike-ins may not experience
the same technical noise as endogenous transcripts <a href="https://www.nature.com/nmeth/journal/v14/n4/full/nmeth.4220.html">(Svensson et al., 2017)</a>.
In addition, scRNASeq experiments often contain only a small number of
spike-ins which reduces our confidence in fitted model parameters.</p>
<div id="highly-variable-genes" class="section level4">
<h4><span class="header-section-number">6.3.1.1</span> Highly Variable Genes</h4>
<p>The first method proposed to identify features in scRNASeq datasets
was to identify highly variable genes (HVG). HVG assumes that if genes
have large differences in expression across cells some of those differences
are due to biological difference between the cells rather than technical noise.
However, because of the nature of count data, there is a positive relationship
between the mean expression of a gene and the variance in the read counts across
cells. This relationship must be corrected for to properly identify HVGs.</p>
<p>For example, you can use the <code>BiocGenerics::rowMeans()</code> and <code>MatrixStats::rowVars()</code>
functions to plot the relationship between mean expression and variance
for all genes in this dataset. Also, you might try using <code>log=&quot;xy&quot;</code> to plot
on a log-scale).</p>
<p>Alternatively, you can use the <code>trendVar()</code> and <code>decomposeVar()</code> functions
in the <code>scran</code> R/Bioconductor package to identify highly variable genes.</p>
<p>For this section, we go back to the <code>tung</code> (or <code>umi.qc</code>) data that
has been normalized and batch corrected. For purposes of the tutorial,
we will select the <code>glm_indi</code> normalized data.</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb157-1" title="1">umi.qc &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/tung/umi_qc.RDS&quot;</span>)</a>
<a class="sourceLine" id="cb157-2" title="2"></a>
<a class="sourceLine" id="cb157-3" title="3">var.fit &lt;-<span class="st"> </span><span class="kw">trendVar</span>(umi.qc, <span class="dt">method =</span> <span class="st">&quot;loess&quot;</span>, <span class="dt">assay.type =</span> <span class="st">&quot;glm_indi&quot;</span>)</a>
<a class="sourceLine" id="cb157-4" title="4">var.out &lt;-<span class="st"> </span><span class="kw">decomposeVar</span>(umi.qc, var.fit)</a>
<a class="sourceLine" id="cb157-5" title="5"><span class="kw">head</span>(var.out)</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 6 columns
##                               mean             total                 bio
##                          &lt;numeric&gt;         &lt;numeric&gt;           &lt;numeric&gt;
## ENSG00000237683   0.24662736578077 0.243858690202493  0.0356801534015803
## ENSG00000187634 0.0358302498678336 0.041042573799668 0.00447500876958429
## ENSG00000188976   1.71485831826354 0.745273783518252   0.114898539528563
## ENSG00000187961  0.218414586619208 0.203530230554452  0.0193960410083338
## ENSG00000187608    1.3692218193445 0.801170759584193   0.142065201109407
## ENSG00000188157   2.02030076886988 0.545656338513709 -0.0424745356608807
##                               tech              p.value
##                          &lt;numeric&gt;            &lt;numeric&gt;
## ENSG00000237683  0.208178536800913  0.00153459146498894
## ENSG00000187634 0.0365675650300837   0.0157736562230252
## ENSG00000188976  0.630375243989689  0.00085174073496591
## ENSG00000187961  0.184134189546118   0.0312433209044445
## ENSG00000187608  0.659105558474786 0.000120336074819414
## ENSG00000188157  0.588130874174589    0.906968345769253
##                                 FDR
##                           &lt;numeric&gt;
## ENSG00000237683  0.0190971382309735
## ENSG00000187634   0.106015932367908
## ENSG00000188976  0.0124082937456012
## ENSG00000187961   0.167734905189692
## ENSG00000187608 0.00271289057563897
## ENSG00000188157                   1</code></pre>
<p>Next, we can plot the mean-variance relationship. The red points are
the control genes (ERCC and MT genes) that are used to estimate the
technical variation. The endogenous genes are plotted in black.</p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb159-1" title="1"><span class="kw">plot</span>(var.out<span class="op">$</span>mean, var.out<span class="op">$</span>total, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">cex=</span><span class="fl">0.6</span>, <span class="dt">xlab=</span><span class="st">&quot;Mean log-expression&quot;</span>,</a>
<a class="sourceLine" id="cb159-2" title="2">    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</a>
<a class="sourceLine" id="cb159-3" title="3"><span class="kw">points</span>(var.fit<span class="op">$</span>mean, var.fit<span class="op">$</span>var, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">pch=</span><span class="dv">16</span>)</a>
<a class="sourceLine" id="cb159-4" title="4">o &lt;-<span class="st"> </span><span class="kw">order</span>(var.out<span class="op">$</span>mean)</a>
<a class="sourceLine" id="cb159-5" title="5"><span class="kw">lines</span>(var.out<span class="op">$</span>mean[o], var.out<span class="op">$</span>tech[o], <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</a></code></pre></div>
<p><img src="06-biological-analyses_files/figure-html/unnamed-chunk-15-1.png" width="90%" style="display: block; margin: auto;" /></p>
<p>Then, the highly variable genes (HVG) are identified as genes
with large positive biological components.</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb160-1" title="1">hvg.out &lt;-<span class="st"> </span>var.out[<span class="kw">which</span>(var.out<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>),]</a>
<a class="sourceLine" id="cb160-2" title="2">hvg.out &lt;-<span class="st"> </span>hvg.out[<span class="kw">order</span>(hvg.out<span class="op">$</span>bio, <span class="dt">decreasing=</span><span class="ot">TRUE</span>),]</a>
<a class="sourceLine" id="cb160-3" title="3"><span class="kw">head</span>(hvg.out)</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 6 columns
##                             mean            total              bio
##                        &lt;numeric&gt;        &lt;numeric&gt;        &lt;numeric&gt;
## ENSG00000106153 2.58010587365124 4.50251391535177 3.99819997557498
## ENSG00000125148 1.10385481967694 2.56632990001896 1.92729406297511
## ENSG00000198865 2.43296165351555 2.36687881721125 1.84095133880717
## ENSG00000110713 3.07884927706287 2.20353806082695 1.76355959764117
## ENSG00000022556 1.87087512946706 2.20770425606636 1.60139777768658
## ENSG00000177105 1.45436561173197 2.03125271629969 1.36969341922523
##                              tech               p.value
##                         &lt;numeric&gt;             &lt;numeric&gt;
## ENSG00000106153 0.504313939776785                     0
## ENSG00000125148 0.639035837043843 1.94361011485433e-234
## ENSG00000198865 0.525927478404086 2.74586525446619e-287
## ENSG00000110713  0.43997846318578                     0
## ENSG00000022556 0.606306478379779 5.88519208672335e-195
## ENSG00000177105 0.661559297074457  7.9450415594479e-138
##                                   FDR
##                             &lt;numeric&gt;
## ENSG00000106153                     0
## ENSG00000125148 5.44210832159212e-231
## ENSG00000198865 1.28140378541756e-283
## ENSG00000110713                     0
## ENSG00000022556 1.17703841734467e-191
## ENSG00000177105 9.26921515268922e-135</code></pre>
<p>These are most highly variable genes after normalization and batch correction.</p>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb162-1" title="1"><span class="kw">nrow</span>(hvg.out)</a></code></pre></div>
<pre><code>## [1] 1561</code></pre>
<p>We can check the distribution of expression values for the top 10 HVGs
to ensure that they are not being driven by outliers. Here are the
violin plots of normalized log-expression values for the top 10 HVGs
in the brain dataset.</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb164-1" title="1"><span class="kw">plotExpression</span>(umi.qc, <span class="kw">rownames</span>(hvg.out)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>], <span class="dt">jitter=</span><span class="st">&quot;jitter&quot;</span>, </a>
<a class="sourceLine" id="cb164-2" title="2">               <span class="dt">exprs_values =</span> <span class="st">&quot;glm_indi&quot;</span>)</a></code></pre></div>
<p><img src="06-biological-analyses_files/figure-html/unnamed-chunk-18-1.png" width="90%" style="display: block; margin: auto;" /></p>
<p>Another approach was proposed by
<a href="http://www.nature.com/nmeth/journal/v10/n11/full/nmeth.2645.html">Brennecke et al.</a>.
To use the Brennecke method, we first normalize for library size then calculate
the mean and the square coefficient of variation (variation divided by
the squared mean expression). A quadratic curve is fit to the relationship
between these two variables for the ERCC spike-in, and then a chi-square
test is used to find genes significantly above the curve.</p>
</div>
<div id="high-dropout-genes" class="section level4">
<h4><span class="header-section-number">6.3.1.2</span> High Dropout Genes</h4>
<p>An alternative to finding HVGs is to identify genes with
unexpectedly high numbers of zeros. The frequency of zeros, known
as the “dropout rate”, is very closely related to expression level
in scRNA-seq data. Zeros are the dominant feature of scRNA-seq data,
typically accounting for over half of the entries in the final expression
matrix. These zeros predominantly result from the failure of mRNAs
failing to be reversed transcribed <a href="http://www.biorxiv.org/content/early/2017/05/25/065094">(Andrews and Hemberg, 2016)</a>.</p>
</div>
</div>
<div id="correlated-expression" class="section level3">
<h3><span class="header-section-number">6.3.2</span> Correlated Expression</h3>
<p>A completely different approach to feature selection is to use
gene-gene correlations. This method is based on the idea that multiple
genes will be differentially expressed between different cell-types
or cell-states. Genes which are expressed in the same cell-population
will be positively correlated with each other where as genes expressed
in different cell-populations will be negatively correated with
each other. Thus important genes can be identified by the magnitude
of their correlation with other genes.</p>
<p>The limitation of this method is that it assumes technical noise is
random and independent for each cell, thus shouldn’t produce gene-gene
correlations, but this assumption is violated by batch effects which are
generally systematic between different experimental batches and will
produce gene-gene correlations. As a result it is more appropriate to
take the top few thousand genes as ranked by gene-gene correlation than
consider the significance of the correlations.</p>
<p>Lastly, another common method for feature selection in scRNA-seq data
is to use PCA loadings. Genes with high PCA loadings are likely to be
highly variable and correlated with many other variable genes, thus
may be relevant to the underlying biology. However, as with gene-gene
correlations PCA loadings tend to be susceptible to detecting systematic
variation due to batch effects; thus it is recommended to plot the PCA
results to determine those components corresponding to the biological
variation rather than batch effects.</p>
</div>
</div>
<div id="differential-expression-de-analysis" class="section level2">
<h2><span class="header-section-number">6.4</span> Differential Expression (DE) analysis</h2>
<div id="bulk-rna-seq-1" class="section level3">
<h3><span class="header-section-number">6.4.1</span> Bulk RNA-seq</h3>
<p>One of the most common types of analyses when working with bulk RNA-seq
data is to identify differentially expressed genes. By comparing the
genes that change between two conditions, e.g. mutant and wild-type or
stimulated and unstimulated, it is possible to characterize the
molecular mechanisms underlying the change.</p>
<p>Several different methods,
e.g. <a href="https://bioconductor.org/packages/DESeq2">DESeq2</a> and
<a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>,
have been developed for bulk RNA-seq. Moreover, there are also
extensive
<a href="http://genomebiology.biomedcentral.com/articles/10.1186/gb-2013-14-9-r95">datasets</a>
available where the RNA-seq data has been validated using
RT-qPCR. These data can be used to benchmark DE finding
algorithms and the available evidence suggests that the
algorithms are performing quite well.</p>
</div>
<div id="single-cell-rna-seq" class="section level3">
<h3><span class="header-section-number">6.4.2</span> Single-cell RNA-seq</h3>
<p>In contrast to bulk RNA-seq, in scRNA-seq we usually do
not have a defined set of experimental conditions. Instead,
as was shown in a previous section we can identify the cell
groups by using an unsupervised clustering approach. Once the
groups have been identified one can find differentially
expressed genes either by comparing the differences in variance
between the groups (like the Kruskal-Wallis test implemented in
SC3), or by comparing gene expression between clusters in a
pairwise manner. In the following section we will mainly
consider tools developed for pairwise comparisons.</p>
</div>
<div id="differences-in-distribution" class="section level3">
<h3><span class="header-section-number">6.4.3</span> Differences in distribution</h3>
<p>Unlike bulk RNA-seq, we generally have a large number of
samples (i.e. cells) for each group we are comparing in single-cell
experiments. Thus we can take advantage of the whole distribution
of expression values in each group to identify differences
between groups rather than only comparing estimates of
mean-expression as is standard for bulk RNA-seq.</p>
<p>There are two main approaches to comparing distributions.
Firstly, we can use existing statistical models/distributions
and fit the same type of model to the expression in each
group then test for differences in the parameters for each
model, or test whether the model fits better if a particular
paramter is allowed to be different according to group. For
instance in section on dealing with confounders, we used
<code>edgeR</code> to test whether allowing mean expression to be different
in different batches significantly improved the fit of a
negative binomial model of the data.</p>
<p>Alternatively, we can use a non-parametric test which does not
assume that expression values follow any particular distribution,
e.g. the <a href="https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test">Kolmogorov-Smirnov test (KS-test)</a>.
Non-parametric tests generally convert observed expression
values to ranks and test whether the distribution of ranks
for one group are signficantly different from the distribution
of ranks for the other group. However, some non-parametric
methods fail in the presence of a large number of tied values,
such as the case for dropouts (zeros) in single-cell RNA-seq
expression data. Moreover, if the conditions for a parametric
test hold, then it will typically be more powerful than a
non-parametric test.</p>
</div>
<div id="models-of-single-cell-rna-seq-data" class="section level3">
<h3><span class="header-section-number">6.4.4</span> Models of single-cell RNA-seq data</h3>
<p>The most common model of RNASeq data is the negative binomial model:</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb165-1" title="1"><span class="kw">set.seed</span>(<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb165-2" title="2"><span class="kw">hist</span>(</a>
<a class="sourceLine" id="cb165-3" title="3">    <span class="kw">rnbinom</span>(</a>
<a class="sourceLine" id="cb165-4" title="4">        <span class="dv">1000</span>, </a>
<a class="sourceLine" id="cb165-5" title="5">        <span class="dt">mu =</span> <span class="dv">10</span>, </a>
<a class="sourceLine" id="cb165-6" title="6">        <span class="dt">size =</span> <span class="dv">100</span>), </a>
<a class="sourceLine" id="cb165-7" title="7">    <span class="dt">col =</span> <span class="st">&quot;grey50&quot;</span>, </a>
<a class="sourceLine" id="cb165-8" title="8">    <span class="dt">xlab =</span> <span class="st">&quot;Read Counts&quot;</span>, </a>
<a class="sourceLine" id="cb165-9" title="9">    <span class="dt">main =</span> <span class="st">&quot;Negative Binomial&quot;</span></a>
<a class="sourceLine" id="cb165-10" title="10">)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:nb-plot"></span>
<img src="06-biological-analyses_files/figure-html/nb-plot-1.png" alt="Negative Binomial distribution of read counts for a single gene across 1000 cells" width="90%" />
<p class="caption">
Figure 6.7: Negative Binomial distribution of read counts for a single gene across 1000 cells
</p>
</div>
<p>Mean:
<span class="math inline">\(\mu = mu\)</span></p>
<p>Variance:
<span class="math inline">\(\sigma^2 = mu + mu^2/size\)</span></p>
<p>It is parameterized by the mean expression (mu) and the dispersion (size),
which is inversely related to the variance. The negative binomial model
fits bulk RNA-seq data very well and it is used for most statistical
methods designed for such data. In addition, it has been show to fit
the distribution of molecule counts obtained from data tagged by unique
molecular identifiers (UMIs) quite well
(<a href="http://www.nature.com/nmeth/journal/v11/n6/full/nmeth.2930.html">Grun et al. 2014</a>,
<a href="http://genome.cshlp.org/content/21/7/1160">Islam et al. 2011</a>).</p>
<p>However, a raw negative binomial model does not fit full-length
transcript data as well due to the high dropout rates relative
to the non-zero read counts. For this type of data a variety of
zero-inflated negative binomial models have been proposed (e.g.
<a href="https://bioconductor.org/packages/release/bioc/html/MAST.html">MAST</a>,
<a href="https://bioconductor.org/packages/release/bioc/html/scde.html">SCDE</a>).</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb166-1" title="1">d &lt;-<span class="st"> </span><span class="fl">0.5</span>;</a>
<a class="sourceLine" id="cb166-2" title="2">counts &lt;-<span class="st"> </span><span class="kw">rnbinom</span>(</a>
<a class="sourceLine" id="cb166-3" title="3">    <span class="dv">1000</span>, </a>
<a class="sourceLine" id="cb166-4" title="4">    <span class="dt">mu =</span> <span class="dv">10</span>, </a>
<a class="sourceLine" id="cb166-5" title="5">    <span class="dt">size =</span> <span class="dv">100</span></a>
<a class="sourceLine" id="cb166-6" title="6">)</a>
<a class="sourceLine" id="cb166-7" title="7">counts[<span class="kw">runif</span>(<span class="dv">1000</span>) <span class="op">&lt;</span><span class="st"> </span>d] &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb166-8" title="8"><span class="kw">hist</span>(</a>
<a class="sourceLine" id="cb166-9" title="9">    counts, </a>
<a class="sourceLine" id="cb166-10" title="10">    <span class="dt">col =</span> <span class="st">&quot;grey50&quot;</span>, </a>
<a class="sourceLine" id="cb166-11" title="11">    <span class="dt">xlab =</span> <span class="st">&quot;Read Counts&quot;</span>, </a>
<a class="sourceLine" id="cb166-12" title="12">    <span class="dt">main =</span> <span class="st">&quot;Zero-inflated NB&quot;</span></a>
<a class="sourceLine" id="cb166-13" title="13">)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:zero-inflation-plot"></span>
<img src="06-biological-analyses_files/figure-html/zero-inflation-plot-1.png" alt="Zero-inflated Negative Binomial distribution" width="90%" />
<p class="caption">
Figure 6.8: Zero-inflated Negative Binomial distribution
</p>
</div>
<p>Mean:
<span class="math inline">\(\mu = mu \cdot (1 - d)\)</span></p>
<p>Variance:
<span class="math inline">\(\sigma^2 = \mu \cdot (1-d) \cdot (1 + d \cdot \mu + \mu / size)\)</span></p>
<p>These models introduce a new parameter <span class="math inline">\(d\)</span>, for the dropout rate,
to the negative binomial model. Turns out, the dropout rate of a gene
is strongly correlated with the mean expression of the gene. Different
zero-inflated negative binomial models use different relationships
between <span class="math inline">\(\mu\)</span> and <span class="math inline">\(d\)</span> and some may fit <span class="math inline">\(\mu\)</span> and <span class="math inline">\(d\)</span> to the expression
of each gene independently.</p>
</div>
</div>
<div id="de-in-tung-dataset" class="section level2">
<h2><span class="header-section-number">6.5</span> DE in <code>tung</code> dataset</h2>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb167-1" title="1"><span class="kw">library</span>(scRNA.seq.funcs)</a>
<a class="sourceLine" id="cb167-2" title="2"><span class="kw">library</span>(edgeR)</a>
<a class="sourceLine" id="cb167-3" title="3"><span class="kw">library</span>(ROCR)</a></code></pre></div>
<div id="introduction-6" class="section level3">
<h3><span class="header-section-number">6.5.1</span> Introduction</h3>
<p>To test different single-cell differential expression methods,
we will continue using the <code>tung</code> dataset. For this experiment
bulk RNA-seq data for each cell-line was generated in addition
to single-cell data. We will use the differentially expressed genes
identified using standard methods on the respective bulk data
as the ground truth for evaluating the accuracy of each single-cell
method. To save time we have pre-computed these for you. You can
run the commands below to load these data.</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb168-1" title="1">DE &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/tung/TPs.txt&quot;</span>)</a>
<a class="sourceLine" id="cb168-2" title="2">notDE &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/tung/TNs.txt&quot;</span>)</a>
<a class="sourceLine" id="cb168-3" title="3">GroundTruth &lt;-<span class="st"> </span><span class="kw">list</span>(</a>
<a class="sourceLine" id="cb168-4" title="4">    <span class="dt">DE =</span> <span class="kw">as.character</span>(<span class="kw">unlist</span>(DE)), </a>
<a class="sourceLine" id="cb168-5" title="5">    <span class="dt">notDE =</span> <span class="kw">as.character</span>(<span class="kw">unlist</span>(notDE))</a>
<a class="sourceLine" id="cb168-6" title="6">)</a></code></pre></div>
<p>This ground truth has been produce for the comparison of
individual <code>NA19101</code> to <code>NA19239</code>.</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb169-1" title="1">umi.qc.sub &lt;-<span class="st"> </span>umi.qc[, umi.qc<span class="op">$</span>individual <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;NA19101&quot;</span>, <span class="st">&quot;NA19239&quot;</span>)]</a>
<a class="sourceLine" id="cb169-2" title="2">group &lt;-<span class="st"> </span><span class="kw">colData</span>(umi.qc.sub)<span class="op">$</span>individual</a>
<a class="sourceLine" id="cb169-3" title="3">batch &lt;-<span class="st"> </span><span class="kw">colData</span>(umi.qc.sub)<span class="op">$</span>batch</a>
<a class="sourceLine" id="cb169-4" title="4">norm_data &lt;-<span class="st"> </span><span class="kw">assay</span>(umi.qc.sub, <span class="st">&quot;glm_indi&quot;</span>)</a></code></pre></div>
<p>Now we will compare various single-cell DE methods.
Note that we will only be running methods which are
available as R-packages and run relatively quickly.</p>
</div>
<div id="kolmogorov-smirnov-test" class="section level3">
<h3><span class="header-section-number">6.5.2</span> Kolmogorov-Smirnov test</h3>
<p>The types of test that are easiest to work with are
non-parametric ones. The most commonly used non-parametric
test is the
<a href="https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test">Kolmogorov-Smirnov test</a>
(KS-test) and we can use it to compare the distributions
for each gene in the two individuals.</p>
<p>The KS-test quantifies the distance between the empirical
cummulative distributions of the expression of each gene
in each of the two populations. It is sensitive to changes
in mean experession and changes in variability. However
it assumes data is continuous and may perform poorly
when data contains a large number of identical values
(e.g. zeros). Another issue with the KS-test is that it can
be very sensitive for large sample sizes and thus it may
end up as significant even though the magnitude of the
difference is very small.</p>
<div class="figure" style="text-align: center"><span id="fig:ks-statistic"></span>
<img src="figures/KS2_Example.png" alt="Illustration of the two-sample Kolmogorov–Smirnov statistic. Red and blue lines each correspond to an empirical distribution function, and the black arrow is the two-sample KS statistic. (taken from [here](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test))" width="60%" />
<p class="caption">
Figure 6.9: Illustration of the two-sample Kolmogorov–Smirnov statistic. Red and blue lines each correspond to an empirical distribution function, and the black arrow is the two-sample KS statistic. (taken from <a href="https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test">here</a>)
</p>
</div>
<p>Now run the test:</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb170-1" title="1">pVals &lt;-<span class="st"> </span><span class="kw">apply</span>(</a>
<a class="sourceLine" id="cb170-2" title="2">    norm_data, <span class="dv">1</span>, <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb170-3" title="3">        <span class="kw">ks.test</span>(</a>
<a class="sourceLine" id="cb170-4" title="4">            x[group <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span>], </a>
<a class="sourceLine" id="cb170-5" title="5">            x[group <span class="op">==</span><span class="st"> &quot;NA19239&quot;</span>]</a>
<a class="sourceLine" id="cb170-6" title="6">        )<span class="op">$</span>p.value</a>
<a class="sourceLine" id="cb170-7" title="7">    }</a>
<a class="sourceLine" id="cb170-8" title="8">)</a>
<a class="sourceLine" id="cb170-9" title="9"><span class="co"># multiple testing correction</span></a>
<a class="sourceLine" id="cb170-10" title="10">pVals &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(pVals, <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</a></code></pre></div>
<p>This code “applies” the function to each row (specified by 1)
of the expression matrix, data. In the function, we are
returning just the <code>p.value</code> from the <code>ks.test()</code> output.
We can now consider how many of the ground truth positive
and negative DE genes are detected by the KS-test.</p>
<div id="evaluating-accuracy" class="section level4">
<h4><span class="header-section-number">6.5.2.1</span> Evaluating Accuracy</h4>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb171-1" title="1">sigDE &lt;-<span class="st"> </span><span class="kw">names</span>(pVals)[pVals <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>]</a>
<a class="sourceLine" id="cb171-2" title="2"></a>
<a class="sourceLine" id="cb171-3" title="3"><span class="co"># Number of KS-DE genes</span></a>
<a class="sourceLine" id="cb171-4" title="4"><span class="kw">length</span>(sigDE) </a></code></pre></div>
<pre><code>## [1] 11942</code></pre>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb173-1" title="1"><span class="co"># Number of KS-DE genes that are true DE genes</span></a>
<a class="sourceLine" id="cb173-2" title="2"><span class="kw">sum</span>(GroundTruth<span class="op">$</span>DE <span class="op">%in%</span><span class="st"> </span>sigDE) </a></code></pre></div>
<pre><code>## [1] 989</code></pre>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb175-1" title="1"><span class="co"># Number of KS-DE genes that are truly not-DE</span></a>
<a class="sourceLine" id="cb175-2" title="2"><span class="kw">sum</span>(GroundTruth<span class="op">$</span>notDE <span class="op">%in%</span><span class="st"> </span>sigDE)</a></code></pre></div>
<pre><code>## [1] 8527</code></pre>
<p>As you can see many more of our ground truth negative genes
were identified as DE by the KS-test (false positives) than
ground truth positive genes (true positives), however this
may be due to the larger number of notDE genes thus we typically
normalize these counts as the True positive rate (TPR), TP/(TP + FN),
and False positive rate (FPR), FP/(FP+TP).</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb177-1" title="1">tp &lt;-<span class="st"> </span><span class="kw">sum</span>(GroundTruth<span class="op">$</span>DE <span class="op">%in%</span><span class="st"> </span>sigDE)</a>
<a class="sourceLine" id="cb177-2" title="2">fp &lt;-<span class="st"> </span><span class="kw">sum</span>(GroundTruth<span class="op">$</span>notDE <span class="op">%in%</span><span class="st"> </span>sigDE)</a>
<a class="sourceLine" id="cb177-3" title="3">tn &lt;-<span class="st"> </span><span class="kw">sum</span>(GroundTruth<span class="op">$</span>notDE <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(pVals)[pVals <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.05</span>])</a>
<a class="sourceLine" id="cb177-4" title="4">fn &lt;-<span class="st"> </span><span class="kw">sum</span>(GroundTruth<span class="op">$</span>DE <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(pVals)[pVals <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.05</span>])</a>
<a class="sourceLine" id="cb177-5" title="5">tpr &lt;-<span class="st"> </span>tp<span class="op">/</span>(tp <span class="op">+</span><span class="st"> </span>fn)</a>
<a class="sourceLine" id="cb177-6" title="6">fpr &lt;-<span class="st"> </span>fp<span class="op">/</span>(fp <span class="op">+</span><span class="st"> </span>tn)</a>
<a class="sourceLine" id="cb177-7" title="7"><span class="kw">cat</span>(<span class="kw">c</span>(tpr, fpr))</a></code></pre></div>
<pre><code>## 0.9347826 0.8212463</code></pre>
<p>Now we can see the TPR is much higher than the FPR
indicating the KS test is identifying DE genes.</p>
<p>So far we’ve only evaluated the performance at a single
significance threshold. Often it is informative to vary
the threshold and evaluate performance across a range of
values. This is then plotted as a receiver-operating-characteristic
curve (ROC) and a general accuracy statistic can be calculated
as the area under this curve (AUC). We will use the <code>ROCR</code> package
to facilitate this plotting.</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb179-1" title="1"><span class="co"># Only consider genes for which we know the ground truth</span></a>
<a class="sourceLine" id="cb179-2" title="2">pVals &lt;-<span class="st"> </span>pVals[<span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>DE <span class="op">|</span><span class="st"> </span></a>
<a class="sourceLine" id="cb179-3" title="3"><span class="st">               </span><span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>notDE] </a>
<a class="sourceLine" id="cb179-4" title="4">truth &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">1</span>, <span class="dt">times =</span> <span class="kw">length</span>(pVals));</a>
<a class="sourceLine" id="cb179-5" title="5">truth[<span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>DE] =<span class="st"> </span><span class="dv">0</span>;</a>
<a class="sourceLine" id="cb179-6" title="6">pred &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">prediction</span>(pVals, truth)</a>
<a class="sourceLine" id="cb179-7" title="7">perf &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">performance</span>(pred, <span class="st">&quot;tpr&quot;</span>, <span class="st">&quot;fpr&quot;</span>)</a>
<a class="sourceLine" id="cb179-8" title="8">ROCR<span class="op">::</span><span class="kw">plot</span>(perf)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ks-roc-plot"></span>
<img src="06-biological-analyses_files/figure-html/ks-roc-plot-1.png" alt="ROC curve for KS-test." width="90%" />
<p class="caption">
Figure 6.10: ROC curve for KS-test.
</p>
</div>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb180-1" title="1">aucObj &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">performance</span>(pred, <span class="st">&quot;auc&quot;</span>)</a>
<a class="sourceLine" id="cb180-2" title="2">aucObj<span class="op">@</span>y.values[[<span class="dv">1</span>]] <span class="co"># AUC</span></a></code></pre></div>
<pre><code>## [1] 0.6717898</code></pre>
<p>Finally, to facilitate the comparisons of other DE methods,
let’s put this code into a function so we don’t need to repeat it:</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb182-1" title="1">DE_Quality_AUC &lt;-<span class="st"> </span><span class="cf">function</span>(pVals) {</a>
<a class="sourceLine" id="cb182-2" title="2">    pVals &lt;-<span class="st"> </span>pVals[<span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>DE <span class="op">|</span><span class="st"> </span></a>
<a class="sourceLine" id="cb182-3" title="3"><span class="st">                   </span><span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>notDE]</a>
<a class="sourceLine" id="cb182-4" title="4">    truth &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">1</span>, <span class="dt">times =</span> <span class="kw">length</span>(pVals));</a>
<a class="sourceLine" id="cb182-5" title="5">    truth[<span class="kw">names</span>(pVals) <span class="op">%in%</span><span class="st"> </span>GroundTruth<span class="op">$</span>DE] =<span class="st"> </span><span class="dv">0</span>;</a>
<a class="sourceLine" id="cb182-6" title="6">    pred &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">prediction</span>(pVals, truth)</a>
<a class="sourceLine" id="cb182-7" title="7">    perf &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">performance</span>(pred, <span class="st">&quot;tpr&quot;</span>, <span class="st">&quot;fpr&quot;</span>)</a>
<a class="sourceLine" id="cb182-8" title="8">    ROCR<span class="op">::</span><span class="kw">plot</span>(perf)</a>
<a class="sourceLine" id="cb182-9" title="9">    aucObj &lt;-<span class="st"> </span>ROCR<span class="op">::</span><span class="kw">performance</span>(pred, <span class="st">&quot;auc&quot;</span>)</a>
<a class="sourceLine" id="cb182-10" title="10">    <span class="kw">return</span>(aucObj<span class="op">@</span>y.values[[<span class="dv">1</span>]])</a>
<a class="sourceLine" id="cb182-11" title="11">}</a></code></pre></div>
</div>
</div>
<div id="wilcoxmann-whitney-u-test" class="section level3">
<h3><span class="header-section-number">6.5.3</span> Wilcox/Mann-Whitney-U Test</h3>
<p>The Wilcox-rank-sum test is another non-parametric test, but
tests specifically if values in one group are greater/less
than the values in the other group. Thus it is often
considered a test for difference in median expression
between two groups; whereas the KS-test is sensitive
to any change in distribution of expression values.</p>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb183-1" title="1">pVals &lt;-<span class="st"> </span><span class="kw">apply</span>(</a>
<a class="sourceLine" id="cb183-2" title="2">    norm_data, <span class="dv">1</span>, <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb183-3" title="3">        <span class="kw">wilcox.test</span>(</a>
<a class="sourceLine" id="cb183-4" title="4">            x[group <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span>], </a>
<a class="sourceLine" id="cb183-5" title="5">            x[group <span class="op">==</span><span class="st"> &quot;NA19239&quot;</span>]</a>
<a class="sourceLine" id="cb183-6" title="6">        )<span class="op">$</span>p.value</a>
<a class="sourceLine" id="cb183-7" title="7">    }</a>
<a class="sourceLine" id="cb183-8" title="8">)</a>
<a class="sourceLine" id="cb183-9" title="9"><span class="co"># multiple testing correction</span></a>
<a class="sourceLine" id="cb183-10" title="10">pVals &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(pVals, <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</a>
<a class="sourceLine" id="cb183-11" title="11"><span class="kw">DE_Quality_AUC</span>(pVals)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:wilcox-plot"></span>
<img src="06-biological-analyses_files/figure-html/wilcox-plot-1.png" alt="ROC curve for Wilcox test." width="90%" />
<p class="caption">
Figure 6.11: ROC curve for Wilcox test.
</p>
</div>
<pre><code>## [1] 0.6779454</code></pre>
</div>
<div id="edger" class="section level3">
<h3><span class="header-section-number">6.5.4</span> edgeR</h3>
<p>edgeR is based on a negative binomial model of gene expression and
uses a generalized linear model (GLM) framework, the enables us
to include other factors such as batch to the model.</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb185-1" title="1">dge &lt;-<span class="st"> </span><span class="kw">DGEList</span>(</a>
<a class="sourceLine" id="cb185-2" title="2">    <span class="dt">counts =</span> <span class="kw">assay</span>(umi.qc.sub, <span class="st">&quot;counts&quot;</span>), </a>
<a class="sourceLine" id="cb185-3" title="3">    <span class="dt">norm.factors =</span> <span class="kw">rep</span>(<span class="dv">1</span>, <span class="kw">length</span>(<span class="kw">assay</span>(umi.qc.sub, <span class="st">&quot;counts&quot;</span>)[<span class="dv">1</span>,])), </a>
<a class="sourceLine" id="cb185-4" title="4">    <span class="dt">group =</span> group</a>
<a class="sourceLine" id="cb185-5" title="5">)</a>
<a class="sourceLine" id="cb185-6" title="6">group_edgeR &lt;-<span class="st"> </span><span class="kw">factor</span>(group)</a>
<a class="sourceLine" id="cb185-7" title="7">design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>group_edgeR)</a>
<a class="sourceLine" id="cb185-8" title="8">dge &lt;-<span class="st"> </span><span class="kw">estimateDisp</span>(dge, <span class="dt">design =</span> design, <span class="dt">trend.method =</span> <span class="st">&quot;none&quot;</span>)</a>
<a class="sourceLine" id="cb185-9" title="9">fit &lt;-<span class="st"> </span><span class="kw">glmFit</span>(dge, design)</a>
<a class="sourceLine" id="cb185-10" title="10">res &lt;-<span class="st"> </span><span class="kw">glmLRT</span>(fit)</a>
<a class="sourceLine" id="cb185-11" title="11">pVals &lt;-<span class="st"> </span>res<span class="op">$</span>table[,<span class="dv">4</span>]</a>
<a class="sourceLine" id="cb185-12" title="12"><span class="kw">names</span>(pVals) &lt;-<span class="st"> </span><span class="kw">rownames</span>(res<span class="op">$</span>table)</a>
<a class="sourceLine" id="cb185-13" title="13"></a>
<a class="sourceLine" id="cb185-14" title="14">pVals &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(pVals, <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</a>
<a class="sourceLine" id="cb185-15" title="15"><span class="kw">DE_Quality_AUC</span>(pVals)</a></code></pre></div>
</div>
<div id="monocle" class="section level3">
<h3><span class="header-section-number">6.5.5</span> Monocle</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/html/monocle.html">Monocle</a> can use several different models for DE. For count data it recommends the Negative Binomial model (negbinomial.size). For normalized data it recommends log-transforming it then using a normal distribution (gaussianff). Similar to edgeR this method uses a GLM framework so in theory can account for batches, however in practice the model fails for this dataset if batches are included.</p>
</div>
<div id="mast" class="section level3">
<h3><span class="header-section-number">6.5.6</span> MAST</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/html/MAST.html">MAST</a> is based on a zero-inflated negative binomial model. It tests for differential expression using a hurdle model to combine tests of discrete (0 vs not zero) and continuous (non-zero values) aspects of gene expression. Again this uses a linear modelling framework to enable complex models to be considered.</p>
</div>
<div id="scde" class="section level3">
<h3><span class="header-section-number">6.5.7</span> SCDE</h3>
<p><a href="http://hms-dbmi.github.io/scde/">SCDE</a> is the first single-cell specific DE method. It fits a zero-inflated negative binomial model to expression data using Bayesian statistics. The usage below tests for differences in mean expression of individual genes across groups but recent versions include methods to test for differences in mean expression or dispersion of groups of genes, usually representing a pathway.</p>
</div>
<div id="zinbwave-deseq2" class="section level3">
<h3><span class="header-section-number">6.5.8</span> zinbwave + DESeq2</h3>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb186-1" title="1"><span class="kw">library</span>(zinbwave)</a>
<a class="sourceLine" id="cb186-2" title="2"></a>
<a class="sourceLine" id="cb186-3" title="3"><span class="co"># low count filter - at least 25 samples with count of 5 or more</span></a>
<a class="sourceLine" id="cb186-4" title="4">keep &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">counts</span>(umi.qc.sub) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span></a>
<a class="sourceLine" id="cb186-5" title="5"><span class="kw">table</span>(keep)</a>
<a class="sourceLine" id="cb186-6" title="6">zinb &lt;-<span class="st"> </span>umi.qc.sub[keep,]</a>
<a class="sourceLine" id="cb186-7" title="7">zinb<span class="op">$</span>individual &lt;-<span class="st">  </span><span class="kw">colData</span>(umi.qc.sub)<span class="op">$</span>individual</a>
<a class="sourceLine" id="cb186-8" title="8"></a>
<a class="sourceLine" id="cb186-9" title="9"><span class="co"># we need to reorganize the assays in the SumExp from splatter</span></a>
<a class="sourceLine" id="cb186-10" title="10">nms &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;counts&quot;</span>, <span class="kw">setdiff</span>(<span class="kw">assayNames</span>(zinb), <span class="st">&quot;counts&quot;</span>))</a>
<a class="sourceLine" id="cb186-11" title="11"><span class="kw">assays</span>(zinb) &lt;-<span class="st"> </span><span class="kw">assays</span>(zinb)[<span class="st">&quot;counts&quot;</span>] <span class="co"># c(&quot;counts&quot;, &quot;glm_indi&quot;)]</span></a>
<a class="sourceLine" id="cb186-12" title="12"><span class="co"># epsilon setting as recommended by the ZINB-WaVE integration paper</span></a>
<a class="sourceLine" id="cb186-13" title="13"><span class="co"># system.time({</span></a>
<a class="sourceLine" id="cb186-14" title="14">  zinb &lt;-<span class="st"> </span><span class="kw">zinbwave</span>(zinb, <span class="dt">K=</span><span class="dv">0</span>, <span class="dt">BPPARAM=</span><span class="kw">SerialParam</span>(), <span class="dt">epsilon=</span><span class="fl">1e12</span>)</a>
<a class="sourceLine" id="cb186-15" title="15"><span class="co"># })</span></a></code></pre></div>
<p>Van den Berge and Perraudeau and others have shown the LRT may perform
better for null hypothesis testing, so we use the LRT. In order to use
the Wald test, it is recommended to set <code>useT=TRUE</code>.</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb187-1" title="1"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(DESeq2))</a>
<a class="sourceLine" id="cb187-2" title="2">dds &lt;-<span class="st"> </span><span class="kw">DESeqDataSet</span>(zinb, <span class="dt">design=</span><span class="op">~</span>individual)</a>
<a class="sourceLine" id="cb187-3" title="3">dds &lt;-<span class="st"> </span><span class="kw">DESeq</span>(dds, <span class="dt">test=</span><span class="st">&quot;LRT&quot;</span>, <span class="dt">reduced=</span><span class="op">~</span><span class="dv">1</span>,</a>
<a class="sourceLine" id="cb187-4" title="4">               <span class="dt">sfType=</span><span class="st">&quot;poscounts&quot;</span>, <span class="dt">minmu=</span><span class="fl">1e-6</span>, <span class="dt">minRep=</span><span class="ot">Inf</span>)</a></code></pre></div>

<div id="refs" class="references">
<div>
<p>Anders, Simon, and Wolfgang Huber. 2010. “Differential Expression Analysis for Sequence Count Data.” <em>Genome Biol</em> 11 (10): R106. <a href="https://doi.org/10.1186/gb-2010-11-10-r106">https://doi.org/10.1186/gb-2010-11-10-r106</a>.</p>
</div>
<div>
<p>Archer, Nathan, Mark D. Walsh, Vahid Shahrezaei, and Daniel Hebenstreit. 2016. “Modeling Enzyme Processivity Reveals That RNA-Seq Libraries Are Biased in Characteristic and Correctable Ways.” <em>Cell Systems</em> 3 (5): 467–479.e12. <a href="https://doi.org/10.1016/j.cels.2016.10.012">https://doi.org/10.1016/j.cels.2016.10.012</a>.</p>
</div>
<div>
<p>Bullard, James H, Elizabeth Purdom, Kasper D Hansen, and Sandrine Dudoit. 2010. “Evaluation of Statistical Methods for Normalization and Differential Expression in mRNA-Seq Experiments.” <em>BMC Bioinformatics</em> 11 (1): 94. <a href="https://doi.org/10.1186/1471-2105-11-94">https://doi.org/10.1186/1471-2105-11-94</a>.</p>
</div>
<div>
<p>Buttner, Maren, Zhichao Miao, Alexander Wolf, Sarah A Teichmann, and Fabian J Theis. 2017. “Assessment of Batch-Correction Methods for scRNA-seq Data with a New Test Metric.” <em>bioRxiv</em>, October, 200345.</p>
</div>
<div>
<p>Deng, Q., D. Ramskold, B. Reinius, and R. Sandberg. 2014. “Single-Cell RNA-Seq Reveals Dynamic, Random Monoallelic Gene Expression in Mammalian Cells.” <em>Science</em> 343 (6167): 193–96. <a href="https://doi.org/10.1126/science.1245316">https://doi.org/10.1126/science.1245316</a>.</p>
</div>
<div>
<p>Gierahn, Todd M, Marc H Wadsworth 2nd, Travis K Hughes, Bryan D Bryson, Andrew Butler, Rahul Satija, Sarah Fortune, J Christopher Love, and Alex K Shalek. 2017. “Seq-Well: Portable, Low-Cost RNA Sequencing of Single Cells at High Throughput.” <em>Nat. Methods</em> 14 (4): 395–98.</p>
</div>
<div>
<p>Guo, Minzhe, Hui Wang, S. Steven Potter, Jeffrey A. Whitsett, and Yan Xu. 2015. “SINCERA: A Pipeline for Single-Cell RNA-Seq Profiling Analysis.” <em>PLoS Comput Biol</em> 11 (11): e1004575. <a href="https://doi.org/10.1371/journal.pcbi.1004575">https://doi.org/10.1371/journal.pcbi.1004575</a>.</p>
</div>
<div>
<p>Haghverdi, Laleh, Aaron T L Lun, Michael D Morgan, and John C Marioni. 2017. “Correcting Batch Effects in Single-Cell RNA Sequencing Data by Matching Mutual Nearest Neighbours.” <em>bioRxiv</em>, July, 165118.</p>
</div>
<div>
<p>Hashimshony, Tamar, Naftalie Senderovich, Gal Avital, Agnes Klochendler, Yaron de Leeuw, Leon Anavy, Dave Gennert, et al. 2016. “CEL-Seq2: Sensitive Highly-Multiplexed Single-Cell RNA-Seq.” <em>Genome Biol</em> 17 (1). <a href="https://doi.org/10.1186/s13059-016-0938-8">https://doi.org/10.1186/s13059-016-0938-8</a>.</p>
</div>
<div>
<p>Hashimshony, Tamar, Florian Wagner, Noa Sher, and Itai Yanai. 2012. “CEL-Seq: Single-cell RNA-Seq by Multiplexed Linear Amplification.” <em>Cell Reports</em> 2 (3): 666–73. <a href="https://doi.org/10.1016/j.celrep.2012.08.003">https://doi.org/10.1016/j.celrep.2012.08.003</a>.</p>
</div>
<div>
<p>Hicks, Stephanie C, F William Townes, Mingxiang Teng, and Rafael A Irizarry. 2018. “Missing Data and Technical Variability in Single-Cell Rna-Sequencing Experiments.” <em>Biostatistics</em> 19 (4): 562–78. <a href="https://doi.org/10.1093/biostatistics/kxx053">https://doi.org/10.1093/biostatistics/kxx053</a>.</p>
</div>
<div>
<p>Islam, Saiful, Amit Zeisel, Simon Joost, La MannoGioele, Pawel Zajac, Maria Kasper, Peter Lönnerberg, and Sten Linnarsson. 2013. “Quantitative Single-Cell RNA-Seq with Unique Molecular Identifiers.” <em>Nat Meth</em> 11 (2): 163–66. <a href="https://doi.org/10.1038/nmeth.2772">https://doi.org/10.1038/nmeth.2772</a>.</p>
</div>
<div>
<p>Jaitin, Diego Adhemar, Ephraim Kenigsberg, Hadas Keren-Shaul, Naama Elefant, Franziska Paul, Irina Zaretsky, Alexander Mildner, et al. 2014. “Massively Parallel Single-Cell RNA-seq for Marker-Free Decomposition of Tissues into Cell Types.” <em>Science</em> 343 (6172): 776–79.</p>
</div>
<div>
<p>Kharchenko, Peter V, Lev Silberstein, and David T Scadden. 2014. “Bayesian Approach to Single-Cell Differential Expression Analysis.” <em>Nat Meth</em> 11 (7): 740–42. <a href="https://doi.org/10.1038/nmeth.2967">https://doi.org/10.1038/nmeth.2967</a>.</p>
</div>
<div>
<p>Kiselev, Vladimir Yu, Kristina Kirschner, Michael T Schaub, Tallulah Andrews, Andrew Yiu, Tamir Chandra, Kedar N Natarajan, et al. 2017. “SC3: Consensus Clustering of Single-Cell RNA-Seq Data.” <em>Nat Meth</em> 14 (5): 483–86. <a href="https://doi.org/10.1038/nmeth.4236">https://doi.org/10.1038/nmeth.4236</a>.</p>
</div>
<div>
<p>Klein, Allon M., Linas Mazutis, Ilke Akartuna, Naren Tallapragada, Adrian Veres, Victor Li, Leonid Peshkin, David A. Weitz, and Marc W. Kirschner. 2015. “Droplet Barcoding for Single-Cell Transcriptomics Applied to Embryonic Stem Cells.” <em>Cell</em> 161 (5): 1187–1201. <a href="https://doi.org/10.1016/j.cell.2015.04.044">https://doi.org/10.1016/j.cell.2015.04.044</a>.</p>
</div>
<div>
<p>Levine, Jacob H., Erin F. Simonds, Sean C. Bendall, Kara L. Davis, El-ad D. Amir, Michelle D. Tadmor, Oren Litvin, et al. 2015. “Data-Driven Phenotypic Dissection of AML Reveals Progenitor-Like Cells That Correlate with Prognosis.” <em>Cell</em> 162 (1): 184–97. <a href="https://doi.org/10.1016/j.cell.2015.05.047">https://doi.org/10.1016/j.cell.2015.05.047</a>.</p>
</div>
<div>
<p>L. Lun, Aaron T., Karsten Bach, and John C. Marioni. 2016. “Pooling Across Cells to Normalize Single-Cell RNA Sequencing Data with Many Zero Counts.” <em>Genome Biol</em> 17 (1). <a href="https://doi.org/10.1186/s13059-016-0947-7">https://doi.org/10.1186/s13059-016-0947-7</a>.</p>
</div>
<div>
<p>Macosko, Evan Z., Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, et al. 2015. “Highly Parallel Genome-Wide Expression Profiling of Individual Cells Using Nanoliter Droplets.” <em>Cell</em> 161 (5): 1202–14. <a href="https://doi.org/10.1016/j.cell.2015.05.002">https://doi.org/10.1016/j.cell.2015.05.002</a>.</p>
</div>
<div>
<p>McCarthy, Davis J., Kieran R. Campbell, Aaron T. L. Lun, and Quin F. Wills. 2017. “Scater: Pre-processing, Quality Control, Normalization and Visualization of Single-Cell RNA-Seq Data in R.” <em>Bioinformatics</em>, January, btw777. <a href="https://doi.org/10.1093/bioinformatics/btw777">https://doi.org/10.1093/bioinformatics/btw777</a>.</p>
</div>
<div>
<p>Picelli, Simone, Åsa K Björklund, Omid R Faridani, Sven Sagasser, Gösta Winberg, and Rickard Sandberg. 2013. “Smart-Seq2 for Sensitive Full-Length Transcriptome Profiling in Single Cells.” <em>Nat Meth</em> 10 (11): 1096–8. <a href="https://doi.org/10.1038/nmeth.2639">https://doi.org/10.1038/nmeth.2639</a>.</p>
</div>
<div>
<p>Picelli, Simone, Omid R Faridani, Asa K Björklund, Gösta Winberg, Sven Sagasser, and Rickard Sandberg. 2014. “Full-Length RNA-seq from Single Cells Using Smart-Seq2.” <em>Nat. Protoc.</em> 9 (1): 171–81.</p>
</div>
<div>
<p>Robinson, Mark D, and Alicia Oshlack. 2010. “A Scaling Normalization Method for Differential Expression Analysis of RNA-Seq Data.” <em>Genome Biol</em> 11 (3): R25. <a href="https://doi.org/10.1186/gb-2010-11-3-r25">https://doi.org/10.1186/gb-2010-11-3-r25</a>.</p>
</div>
<div>
<p>Soumillon, Magali, Davide Cacchiarelli, Stefan Semrau, Alexander van Oudenaarden, and Tarjei S Mikkelsen. 2014. “Characterization of Directed Differentiation by High-Throughput Single-Cell RNA-Seq.” <em>bioRxiv</em>, March, 003236.</p>
</div>
<div>
<p>Stegle, Oliver, Sarah A. Teichmann, and John C. Marioni. 2015. “Computational and Analytical Challenges in Single-Cell Transcriptomics.” <em>Nat Rev Genet</em> 16 (3): 133–45. <a href="https://doi.org/10.1038/nrg3833">https://doi.org/10.1038/nrg3833</a>.</p>
</div>
<div>
<p>Svensson, Valentine, Kedar Nath Natarajan, Lam-Ha Ly, Ricardo J Miragaia, Charlotte Labalette, Iain C Macaulay, Ana Cvejic, and Sarah A Teichmann. 2017. “Power Analysis of Single-Cell RNA-Sequencing Experiments.” <em>Nat Meth</em> 14 (4): 381–87. <a href="https://doi.org/10.1038/nmeth.4220">https://doi.org/10.1038/nmeth.4220</a>.</p>
</div>
<div>
<p>Tang, Fuchou, Catalin Barbacioru, Yangzhou Wang, Ellen Nordman, Clarence Lee, Nanlan Xu, Xiaohui Wang, et al. 2009. “mRNA-Seq Whole-Transcriptome Analysis of a Single Cell.” <em>Nat Meth</em> 6 (5): 377–82. <a href="https://doi.org/10.1038/nmeth.1315">https://doi.org/10.1038/nmeth.1315</a>.</p>
</div>
<div>
<p>Tung, Po-Yuan, John D. Blischak, Chiaowen Joyce Hsiao, David A. Knowles, Jonathan E. Burnett, Jonathan K. Pritchard, and Yoav Gilad. 2017. “Batch Effects and the Effective Design of Single-Cell Gene Expression Studies.” <em>Sci. Rep.</em> 7 (January): 39921. <a href="https://doi.org/10.1038/srep39921">https://doi.org/10.1038/srep39921</a>.</p>
</div>
<div>
<p>Xu, Chen, and Zhengchang Su. 2015. “Identification of Cell Types from Single-Cell Transcriptomes Using a Novel Clustering Method.” <em>Bioinformatics</em> 31 (12): 1974–80. <a href="https://doi.org/10.1093/bioinformatics/btv088">https://doi.org/10.1093/bioinformatics/btv088</a>.</p>
</div>
<div>
<p>Zheng, Grace X Y, Jessica M Terry, Phillip Belgrader, Paul Ryvkin, Zachary W Bent, Ryan Wilson, Solongo B Ziraldo, et al. 2017. “Massively Parallel Digital Transcriptional Profiling of Single Cells.” <em>Nat Commun</em> 8 (January): 14049. <a href="https://doi.org/10.1038/ncomms14049">https://doi.org/10.1038/ncomms14049</a>.</p>
</div>
<div>
<p>Ziegenhain, Christoph, Beate Vieth, Swati Parekh, Björn Reinius, Amy Guillaumet-Adkins, Martha Smets, Heinrich Leonhardt, Holger Heyn, Ines Hellmann, and Wolfgang Enard. 2017. “Comparative Analysis of Single-Cell RNA Sequencing Methods.” <em>Molecular Cell</em> 65 (4): 631–643.e4. <a href="https://doi.org/10.1016/j.molcel.2017.01.023">https://doi.org/10.1016/j.molcel.2017.01.023</a>.</p>
</div>
<div>
<p>žurauskienė, Justina, and Christopher Yau. 2016. “pcaReduce: Hierarchical Clustering of Single Cell Transcriptional Profiles.” <em>BMC Bioinformatics</em> 17 (1). <a href="https://doi.org/10.1186/s12859-016-0984-y">https://doi.org/10.1186/s12859-016-0984-y</a>.</p>
</div>
</div>
</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-Kiselev2016-bq">
<p>Kiselev, Vladimir Yu, Kristina Kirschner, Michael T Schaub, Tallulah Andrews, Andrew Yiu, Tamir Chandra, Kedar N Natarajan, et al. 2017. “SC3: Consensus Clustering of Single-Cell RNA-Seq Data.” <em>Nat Meth</em> 14 (5): 483–86. <a href="https://doi.org/10.1038/nmeth.4236">https://doi.org/10.1038/nmeth.4236</a>.</p>
</div>
<div id="ref-Guo2015-ok">
<p>Guo, Minzhe, Hui Wang, S. Steven Potter, Jeffrey A. Whitsett, and Yan Xu. 2015. “SINCERA: A Pipeline for Single-Cell RNA-Seq Profiling Analysis.” <em>PLoS Comput Biol</em> 11 (11): e1004575. <a href="https://doi.org/10.1371/journal.pcbi.1004575">https://doi.org/10.1371/journal.pcbi.1004575</a>.</p>
</div>
<div id="ref-Zurauskiene2016-kg">
<p>žurauskienė, Justina, and Christopher Yau. 2016. “pcaReduce: Hierarchical Clustering of Single Cell Transcriptional Profiles.” <em>BMC Bioinformatics</em> 17 (1). <a href="https://doi.org/10.1186/s12859-016-0984-y">https://doi.org/10.1186/s12859-016-0984-y</a>.</p>
</div>
<div id="ref-Xu2015-vf">
<p>Xu, Chen, and Zhengchang Su. 2015. “Identification of Cell Types from Single-Cell Transcriptomes Using a Novel Clustering Method.” <em>Bioinformatics</em> 31 (12): 1974–80. <a href="https://doi.org/10.1093/bioinformatics/btv088">https://doi.org/10.1093/bioinformatics/btv088</a>.</p>
</div>
<div id="ref-Levine2015-fk">
<p>Levine, Jacob H., Erin F. Simonds, Sean C. Bendall, Kara L. Davis, El-ad D. Amir, Michelle D. Tadmor, Oren Litvin, et al. 2015. “Data-Driven Phenotypic Dissection of AML Reveals Progenitor-Like Cells That Correlate with Prognosis.” <em>Cell</em> 162 (1): 184–97. <a href="https://doi.org/10.1016/j.cell.2015.05.047">https://doi.org/10.1016/j.cell.2015.05.047</a>.</p>
</div>
<div id="ref-Deng2014-mx">
<p>Deng, Q., D. Ramskold, B. Reinius, and R. Sandberg. 2014. “Single-Cell RNA-Seq Reveals Dynamic, Random Monoallelic Gene Expression in Mammalian Cells.” <em>Science</em> 343 (6167): 193–96. <a href="https://doi.org/10.1126/science.1245316">https://doi.org/10.1126/science.1245316</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="cleaning-the-expression-matrix.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["scRNA-seq-course.pdf"],
"toc": {
"collapse": "section"
},
"search": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
