<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop</title>
  <meta name="description" content="Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop" />
  
  
  

<meta name="author" content="Stephanie Hicks (stephaniehicks)">


<meta name="date" content="2018-12-03">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="construction-of-expression-matrix.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Welcome!</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#preface"><i class="fa fa-check"></i><b>1.1</b> Preface</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#about-the-course"><i class="fa fa-check"></i><b>1.2</b> About the course</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.3</b> GitHub</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.4</b> Installation</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.6</b> Prerequisites</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.7</b> Contact</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html"><i class="fa fa-check"></i><b>2</b> Introduction to single-cell RNA-seq</a><ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#bulk-rna-seq"><i class="fa fa-check"></i><b>2.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#scrna-seq"><i class="fa fa-check"></i><b>2.2</b> scRNA-seq</a></li>
<li class="chapter" data-level="2.3" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#workflow"><i class="fa fa-check"></i><b>2.3</b> Workflow</a></li>
<li class="chapter" data-level="2.4" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#computational-analysis"><i class="fa fa-check"></i><b>2.4</b> Computational Analysis</a></li>
<li class="chapter" data-level="2.5" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#challenges"><i class="fa fa-check"></i><b>2.5</b> Challenges</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#experimental-methods"><i class="fa fa-check"></i><b>2.6</b> Experimental methods</a></li>
<li class="chapter" data-level="2.7" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#what-platform-to-use-for-my-experiment"><i class="fa fa-check"></i><b>2.7</b> What platform to use for my experiment?</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html"><i class="fa fa-check"></i><b>3</b> Construction of expression matrix</a><ul>
<li class="chapter" data-level="3.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#raw-sequencing-reads-qc"><i class="fa fa-check"></i><b>3.1</b> Raw sequencing reads QC</a><ul>
<li class="chapter" data-level="3.1.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#fastq"><i class="fa fa-check"></i><b>3.1.1</b> FastQ</a></li>
<li class="chapter" data-level="3.1.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#check-the-quality-of-the-raw-sequencing-reads"><i class="fa fa-check"></i><b>3.1.2</b> Check the quality of the raw sequencing reads</a></li>
<li class="chapter" data-level="3.1.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#trim-sequencing-adapaters-from-sequencing-reads"><i class="fa fa-check"></i><b>3.1.3</b> Trim sequencing adapaters from sequencing reads</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-alignment"><i class="fa fa-check"></i><b>3.2</b> Reads alignment</a><ul>
<li class="chapter" data-level="3.2.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#genome-fasta-gtf"><i class="fa fa-check"></i><b>3.2.1</b> Genome (<code>FASTA</code>, <code>GTF</code>)</a></li>
<li class="chapter" data-level="3.2.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#alignment-using-star"><i class="fa fa-check"></i><b>3.2.2</b> Alignment using Star</a></li>
<li class="chapter" data-level="3.2.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#pseudo-alignment-using-salmon"><i class="fa fa-check"></i><b>3.2.3</b> Pseudo-alignment using Salmon</a></li>
<li class="chapter" data-level="3.2.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapped-reads-bam-file-format"><i class="fa fa-check"></i><b>3.2.4</b> Mapped reads: <code>BAM</code> file format</a></li>
<li class="chapter" data-level="3.2.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cram"><i class="fa fa-check"></i><b>3.2.5</b> CRAM</a></li>
<li class="chapter" data-level="3.2.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-mapping-qc"><i class="fa fa-check"></i><b>3.2.6</b> Reads Mapping QC</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-quantification"><i class="fa fa-check"></i><b>3.3</b> Reads quantification</a></li>
<li class="chapter" data-level="3.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#unique-molecular-identifiers-umis"><i class="fa fa-check"></i><b>3.4</b> Unique Molecular Identifiers (UMIs)</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html"><i class="fa fa-check"></i><b>4</b> Introduction to R/Bioconductor</a><ul>
<li class="chapter" data-level="4.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#installing-r-packages"><i class="fa fa-check"></i><b>4.1</b> Installing R packages</a><ul>
<li class="chapter" data-level="4.1.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#cran"><i class="fa fa-check"></i><b>4.1.1</b> CRAN</a></li>
<li class="chapter" data-level="4.1.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#github-1"><i class="fa fa-check"></i><b>4.1.2</b> Github</a></li>
<li class="chapter" data-level="4.1.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#bioconductor"><i class="fa fa-check"></i><b>4.1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="4.1.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#source"><i class="fa fa-check"></i><b>4.1.4</b> Source</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#data-types"><i class="fa fa-check"></i><b>4.2</b> Data Types</a><ul>
<li class="chapter" data-level="4.2.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#what-is-tidy-data"><i class="fa fa-check"></i><b>4.2.1</b> What is Tidy Data?</a></li>
<li class="chapter" data-level="4.2.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#what-is-rich-data"><i class="fa fa-check"></i><b>4.2.2</b> What is Rich Data?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#bioconductor-1"><i class="fa fa-check"></i><b>4.3</b> Bioconductor</a><ul>
<li class="chapter" data-level="4.3.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#singlecellexperiment-class"><i class="fa fa-check"></i><b>4.3.1</b> <code>SingleCellExperiment</code> class</a></li>
<li class="chapter" data-level="4.3.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#scater-package"><i class="fa fa-check"></i><b>4.3.2</b> <code>scater</code> package</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#data-visualization-using-ggplot2"><i class="fa fa-check"></i><b>4.4</b> Data visualization using ggplot2</a><ul>
<li class="chapter" data-level="4.4.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#what-is-ggplot2"><i class="fa fa-check"></i><b>4.4.1</b> What is ggplot2?</a></li>
<li class="chapter" data-level="4.4.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#principles-of-ggplot2"><i class="fa fa-check"></i><b>4.4.2</b> Principles of ggplot2</a></li>
<li class="chapter" data-level="4.4.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#using-the-aes-mapping-function"><i class="fa fa-check"></i><b>4.4.3</b> Using the <code>aes()</code> mapping function</a></li>
<li class="chapter" data-level="4.4.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#geoms"><i class="fa fa-check"></i><b>4.4.4</b> Geoms</a></li>
<li class="chapter" data-level="4.4.5" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#plotting-data-from-more-than-2-cells"><i class="fa fa-check"></i><b>4.4.5</b> Plotting data from more than 2 cells</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html"><i class="fa fa-check"></i><b>5</b> Cleaning the expression matrix</a><ul>
<li class="chapter" data-level="5.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#expression-qc"><i class="fa fa-check"></i><b>5.1</b> Expression QC</a><ul>
<li class="chapter" data-level="5.1.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction"><i class="fa fa-check"></i><b>5.1.1</b> Introduction</a></li>
<li class="chapter" data-level="5.1.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#tung-dataset-umi"><i class="fa fa-check"></i><b>5.1.2</b> Tung dataset (UMI)</a></li>
<li class="chapter" data-level="5.1.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cell-qc"><i class="fa fa-check"></i><b>5.1.3</b> Cell QC</a></li>
<li class="chapter" data-level="5.1.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cell-filtering"><i class="fa fa-check"></i><b>5.1.4</b> Cell filtering</a></li>
<li class="chapter" data-level="5.1.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#compare-filterings"><i class="fa fa-check"></i><b>5.1.5</b> Compare filterings</a></li>
<li class="chapter" data-level="5.1.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#gene-analysis"><i class="fa fa-check"></i><b>5.1.6</b> Gene analysis</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#data-visualization"><i class="fa fa-check"></i><b>5.2</b> Data visualization</a><ul>
<li class="chapter" data-level="5.2.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-1"><i class="fa fa-check"></i><b>5.2.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#pca-plot"><i class="fa fa-check"></i><b>5.2.2</b> PCA plot</a></li>
<li class="chapter" data-level="5.2.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#tsne-map"><i class="fa fa-check"></i><b>5.2.3</b> tSNE map</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#identifying-confounding-factors"><i class="fa fa-check"></i><b>5.3</b> Identifying confounding factors</a><ul>
<li class="chapter" data-level="5.3.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-2"><i class="fa fa-check"></i><b>5.3.1</b> Introduction</a></li>
<li class="chapter" data-level="5.3.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#correlations-with-pcs"><i class="fa fa-check"></i><b>5.3.2</b> Correlations with PCs</a></li>
<li class="chapter" data-level="5.3.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#explanatory-variables"><i class="fa fa-check"></i><b>5.3.3</b> Explanatory variables</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalization-theory"><i class="fa fa-check"></i><b>5.4</b> Normalization theory</a><ul>
<li class="chapter" data-level="5.4.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-3"><i class="fa fa-check"></i><b>5.4.1</b> Introduction</a></li>
<li class="chapter" data-level="5.4.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#library-size-1"><i class="fa fa-check"></i><b>5.4.2</b> Library size</a></li>
<li class="chapter" data-level="5.4.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalisations"><i class="fa fa-check"></i><b>5.4.3</b> Normalisations</a></li>
<li class="chapter" data-level="5.4.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#effectiveness"><i class="fa fa-check"></i><b>5.4.4</b> Effectiveness</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalization-practice"><i class="fa fa-check"></i><b>5.5</b> Normalization practice</a><ul>
<li class="chapter" data-level="5.5.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#raw"><i class="fa fa-check"></i><b>5.5.1</b> Raw</a></li>
<li class="chapter" data-level="5.5.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cpm-1"><i class="fa fa-check"></i><b>5.5.2</b> CPM</a></li>
<li class="chapter" data-level="5.5.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#scran-1"><i class="fa fa-check"></i><b>5.5.3</b> scran</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#dealing-with-confounders"><i class="fa fa-check"></i><b>5.6</b> Dealing with confounders</a><ul>
<li class="chapter" data-level="5.6.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-4"><i class="fa fa-check"></i><b>5.6.1</b> Introduction</a></li>
<li class="chapter" data-level="5.6.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#combat"><i class="fa fa-check"></i><b>5.6.2</b> Combat</a></li>
<li class="chapter" data-level="5.6.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#mnncorrect"><i class="fa fa-check"></i><b>5.6.3</b> mnnCorrect</a></li>
<li class="chapter" data-level="5.6.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#glm"><i class="fa fa-check"></i><b>5.6.4</b> GLM</a></li>
<li class="chapter" data-level="5.6.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#how-to-evaluate-and-compare-confounder-removal-strategies"><i class="fa fa-check"></i><b>5.6.5</b> How to evaluate and compare confounder removal strategies</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="biological-analysis.html"><a href="biological-analysis.html"><i class="fa fa-check"></i><b>6</b> Biological Analysis</a><ul>
<li class="chapter" data-level="6.1" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering"><i class="fa fa-check"></i><b>6.1</b> Clustering</a><ul>
<li class="chapter" data-level="6.1.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-5"><i class="fa fa-check"></i><b>6.1.1</b> Introduction</a></li>
<li class="chapter" data-level="6.1.2" data-path="biological-analysis.html"><a href="biological-analysis.html#dimensionality-reduction"><i class="fa fa-check"></i><b>6.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.1.3" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-methods"><i class="fa fa-check"></i><b>6.1.3</b> Clustering methods</a></li>
<li class="chapter" data-level="6.1.4" data-path="biological-analysis.html"><a href="biological-analysis.html#challenges-in-clustering"><i class="fa fa-check"></i><b>6.1.4</b> Challenges in clustering</a></li>
<li class="chapter" data-level="6.1.5" data-path="biological-analysis.html"><a href="biological-analysis.html#software-packages-for-scrna-seq-data"><i class="fa fa-check"></i><b>6.1.5</b> Software packages for scRNA-seq data</a></li>
<li class="chapter" data-level="6.1.6" data-path="biological-analysis.html"><a href="biological-analysis.html#comparing-clustering"><i class="fa fa-check"></i><b>6.1.6</b> Comparing clustering</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-example"><i class="fa fa-check"></i><b>6.2</b> Clustering example</a><ul>
<li class="chapter" data-level="6.2.1" data-path="biological-analysis.html"><a href="biological-analysis.html#deng-dataset"><i class="fa fa-check"></i><b>6.2.1</b> Deng dataset</a></li>
<li class="chapter" data-level="6.2.2" data-path="biological-analysis.html"><a href="biological-analysis.html#sc3-1"><i class="fa fa-check"></i><b>6.2.2</b> SC3</a></li>
<li class="chapter" data-level="6.2.3" data-path="biological-analysis.html"><a href="biological-analysis.html#tsne-kmeans"><i class="fa fa-check"></i><b>6.2.3</b> tSNE + kmeans</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="biological-analysis.html"><a href="biological-analysis.html#feature-selection"><i class="fa fa-check"></i><b>6.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="6.3.1" data-path="biological-analysis.html"><a href="biological-analysis.html#identifying-genes-vs-a-null-model"><i class="fa fa-check"></i><b>6.3.1</b> Identifying Genes vs a Null Model</a></li>
<li class="chapter" data-level="6.3.2" data-path="biological-analysis.html"><a href="biological-analysis.html#correlated-expression"><i class="fa fa-check"></i><b>6.3.2</b> Correlated Expression</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="biological-analysis.html"><a href="biological-analysis.html#differential-expression-de-analysis"><i class="fa fa-check"></i><b>6.4</b> Differential Expression (DE) analysis</a><ul>
<li class="chapter" data-level="6.4.1" data-path="biological-analysis.html"><a href="biological-analysis.html#bulk-rna-seq-1"><i class="fa fa-check"></i><b>6.4.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="6.4.2" data-path="biological-analysis.html"><a href="biological-analysis.html#single-cell-rna-seq"><i class="fa fa-check"></i><b>6.4.2</b> Single-cell RNA-seq</a></li>
<li class="chapter" data-level="6.4.3" data-path="biological-analysis.html"><a href="biological-analysis.html#differences-in-distribution"><i class="fa fa-check"></i><b>6.4.3</b> Differences in distribution</a></li>
<li class="chapter" data-level="6.4.4" data-path="biological-analysis.html"><a href="biological-analysis.html#models-of-single-cell-rna-seq-data"><i class="fa fa-check"></i><b>6.4.4</b> Models of single-cell RNA-seq data</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="biological-analysis.html"><a href="biological-analysis.html#de-in-tung-dataset"><i class="fa fa-check"></i><b>6.5</b> DE in <code>tung</code> dataset</a><ul>
<li class="chapter" data-level="6.5.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-6"><i class="fa fa-check"></i><b>6.5.1</b> Introduction</a></li>
<li class="chapter" data-level="6.5.2" data-path="biological-analysis.html"><a href="biological-analysis.html#kolmogorov-smirnov-test"><i class="fa fa-check"></i><b>6.5.2</b> Kolmogorov-Smirnov test</a></li>
<li class="chapter" data-level="6.5.3" data-path="biological-analysis.html"><a href="biological-analysis.html#wilcoxmann-whitney-u-test"><i class="fa fa-check"></i><b>6.5.3</b> Wilcox/Mann-Whitney-U Test</a></li>
<li class="chapter" data-level="6.5.4" data-path="biological-analysis.html"><a href="biological-analysis.html#edger-or-deseq2"><i class="fa fa-check"></i><b>6.5.4</b> edgeR or DESeq2</a></li>
<li class="chapter" data-level="6.5.5" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle"><i class="fa fa-check"></i><b>6.5.5</b> Monocle</a></li>
<li class="chapter" data-level="6.5.6" data-path="biological-analysis.html"><a href="biological-analysis.html#mast"><i class="fa fa-check"></i><b>6.5.6</b> MAST</a></li>
<li class="chapter" data-level="6.5.7" data-path="biological-analysis.html"><a href="biological-analysis.html#scde"><i class="fa fa-check"></i><b>6.5.7</b> SCDE</a></li>
<li class="chapter" data-level="6.5.8" data-path="biological-analysis.html"><a href="biological-analysis.html#zinbwave-deseq2edger"><i class="fa fa-check"></i><b>6.5.8</b> zinbwave + DESeq2/edger</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="http://www.sanger.ac.uk/science/groups/hemberg-group" target="blank">Hemberg Lab</a></li>
<li><a href="http://www.stephaniehicks.comp" target="blank">Stephanie Hicks</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-single-cell-rna-seq" class="section level1">
<h1><span class="header-section-number">2</span> Introduction to single-cell RNA-seq</h1>
<div id="bulk-rna-seq" class="section level2">
<h2><span class="header-section-number">2.1</span> Bulk RNA-seq</h2>
<ul>
<li>A major breakthrough (replaced microarrays) in the late 2000’s and has been widely used since</li>
<li>Measures the <strong>average expression level</strong> for each gene across a large population of input cells</li>
<li>Useful for comparative transcriptomics, e.g. samples of the same tissue from different species</li>
<li>Useful for quantifying expression signatures from ensembles, e.g. in disease studies</li>
<li><strong>Insufficient</strong> for studying heterogeneous systems, e.g. early development studies, complex tissues (brain)</li>
<li>Does <strong>not</strong> provide insights into the stochastic nature of gene expression</li>
</ul>
</div>
<div id="scrna-seq" class="section level2">
<h2><span class="header-section-number">2.2</span> scRNA-seq</h2>
<ul>
<li>A <strong>new</strong> technology, first publication by <span class="citation">(Tang et al. <a href="#ref-Tang2009-bu">2009</a>)</span></li>
<li>Did not gain widespread popularity until <a href="https://www.ohio.edu/bioinformatics/upload/Single-Cell-RNA-seq-Method-of-the-Year-2013.pdf">~2014</a> when new protocols and lower sequencing costs made it more accessible</li>
<li>Measures the <strong>distribution of expression levels</strong> for each gene across a population of cells</li>
<li>Allows to study new biological questions in which <strong>cell-specific changes in transcriptome are important</strong>, e.g. cell type identification, heterogeneity of cell responses, stochasticity of gene expression, inference of gene regulatory networks across the cells.</li>
<li>Datasets range <strong>from <span class="math inline">\(10^2\)</span> to <span class="math inline">\(10^6\)</span> cells</strong> and increase in size every year</li>
<li>Currently there are several different protocols in use, e.g. SMART-seq2 <span class="citation">(Picelli et al. <a href="#ref-Picelli2013-sb">2013</a>)</span>, CELL-seq <span class="citation">(Hashimshony et al. <a href="#ref-Hashimshony2012-kd">2012</a>)</span> and Drop-seq <span class="citation">(Macosko et al. <a href="#ref-Macosko2015-ix">2015</a>)</span></li>
<li>There are also commercial platforms available, including the <a href="https://www.fluidigm.com/products/c1-system">Fluidigm C1</a>, <a href="https://www.wafergen.com/products/icell8-single-cell-system">Wafergen ICELL8</a> and the <a href="https://www.10xgenomics.com/single-cell/">10X Genomics Chromium</a></li>
<li>Several computational analysis methods from bulk RNA-seq <strong>can</strong> be used</li>
<li><strong>But in most cases</strong> computational analysis requires adaptation of the existing methods or development of new ones</li>
</ul>
</div>
<div id="workflow" class="section level2">
<h2><span class="header-section-number">2.3</span> Workflow</h2>
<div class="figure" style="text-align: center"><span id="fig:intro-rna-seq-workflow"></span>
<img src="figures/RNA-Seq_workflow-5.pdf.jpg" alt="Single cell sequencing (taken from Wikipedia)" width="90%" />
<p class="caption">
Figure 2.1: Single cell sequencing (taken from Wikipedia)
</p>
</div>
<p>Overall, experimental scRNA-seq protocols are similar to the methods used for bulk RNA-seq. We will be discussing some of the most common approaches in the next chapter.</p>
</div>
<div id="computational-analysis" class="section level2">
<h2><span class="header-section-number">2.4</span> Computational Analysis</h2>
<p>This course is concerned with the computational analysis of the data
obtained from scRNA-seq experiments. The first steps (yellow) are general for any highthroughput sequencing data. Later steps (orange) require a mix of existing RNASeq analysis methods and novel methods to address the technical difference of scRNASeq. Finally the biological interpretation (blue) <strong>should</strong> be analyzed with methods specifically developed for scRNASeq.</p>
<div class="figure" style="text-align: center"><span id="fig:intro-flowchart"></span>
<img src="figures/flowchart.png" alt="Flowchart of the scRNA-seq analysis" width="65%" />
<p class="caption">
Figure 2.2: Flowchart of the scRNA-seq analysis
</p>
</div>
<p>There are several reviews of the scRNA-seq analysis available including <span class="citation">(Stegle, Teichmann, and Marioni <a href="#ref-Stegle2015-uv">2015</a>)</span>.</p>
<p>Today, there are also several different analysis tools available for
carrying out one or more steps in the flowchart above. These include:</p>
<ul>
<li><a href="https://master.bioconductor.org/packages/release/workflows/html/simpleSingleCell.html">Bioconductor</a> is a open-source, open-development software project for the analysis of high-throughput genomics data, including packages for the analysis of single-cell data.</li>
<li><a href="http://satijalab.org/seurat/">Seurat</a> is an R package designed for QC, analysis, and exploration of single cell RNA-seq data.</li>
<li><a href="https://scanpy.readthedocs.io/en/stable/">scanpy</a> is a Python-based, scalable toolkit for analyzing single-cell gene expression data. It includes preprocessing, visualization, clustering, pseudotime and trajectory inference and differential expression testing.</li>
<li><a href="https://github.com/VCCRI/Falco/">Falco</a> a single-cell RNA-seq processing framework on the cloud.</li>
<li><a href="https://asap.epfl.ch/">ASAP</a> (Automated Single-cell Analysis Pipeline) is an interactive web-based platform for single-cell analysis.</li>
</ul>
</div>
<div id="challenges" class="section level2">
<h2><span class="header-section-number">2.5</span> Challenges</h2>
<p>The main difference between bulk and single cell RNA-seq is that each sequencing library represents a single cell, instead of a population of cells. Therefore, significant attention has to be paid to comparison of the results from different cells (sequencing libraries). The main sources of discrepancy between the libraries are:</p>
<ul>
<li><strong>Amplification</strong> (up to 1 million fold)</li>
<li><strong>Gene ‘dropouts’</strong> in which a gene is observed at a moderate expression level in one cell but is not detected in another cell <span class="citation">(Kharchenko, Silberstein, and Scadden <a href="#ref-Kharchenko2014-ts">2014</a>, <span class="citation">@Hicks:2018aa</span>)</span>.</li>
</ul>
<p>In both cases the discrepancies are introduced due to low starting amounts of transcripts since the RNA comes from one cell only. Improving the transcript capture efficiency and reducing the amplification bias are currently active areas of research. However, as we shall see in this course, it is possible to alleviate some of these issues through proper normalization and corrections.</p>
</div>
<div id="experimental-methods" class="section level2">
<h2><span class="header-section-number">2.6</span> Experimental methods</h2>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-2"></span>
<img src="figures/moores-law.png" alt="Moore's law in single cell transcriptomics (image taken from [Svensson et al](https://arxiv.org/abs/1704.01379))" width="100%" />
<p class="caption">
Figure 2.3: Moore’s law in single cell transcriptomics (image taken from <a href="https://arxiv.org/abs/1704.01379">Svensson et al</a>)
</p>
</div>
<p>Development of new methods and protocols for scRNA-seq
is currently a very active area of research, and several
protocols have been published over the last few years. A
non-comprehensive list includes:</p>
<ul>
<li>CEL-seq <span class="citation">(Hashimshony et al. <a href="#ref-Hashimshony2012-kd">2012</a>)</span></li>
<li>CEL-seq2 <span class="citation">(Hashimshony et al. <a href="#ref-Hashimshony2016-lx">2016</a>)</span></li>
<li>Drop-seq <span class="citation">(Macosko et al. <a href="#ref-Macosko2015-ix">2015</a>)</span></li>
<li>InDrop-seq <span class="citation">(Klein et al. <a href="#ref-Klein2015-kz">2015</a>)</span></li>
<li>MARS-seq <span class="citation">(Jaitin et al. <a href="#ref-Jaitin2014-ko">2014</a>)</span></li>
<li>SCRB-seq <span class="citation">(Soumillon et al. <a href="#ref-Soumillon2014-eu">2014</a>)</span></li>
<li>Seq-well <span class="citation">(Gierahn et al. <a href="#ref-Gierahn2017-es">2017</a>)</span></li>
<li>Smart-seq <span class="citation">(Picelli et al. <a href="#ref-Picelli2014-ic">2014</a>)</span></li>
<li>Smart-seq2 <span class="citation">(Picelli et al. <a href="#ref-Picelli2014-ic">2014</a>)</span></li>
<li><a href="http://www.clontech.com/US/Products/cDNA_Synthesis_and_Library_Construction/Next_Gen_Sequencing_Kits/Total_RNA-Seq/Universal_RNA_Seq_Random_Primed">SMARTer</a></li>
<li>STRT-seq <span class="citation">(Islam et al. <a href="#ref-Islam2014-cn">2013</a>)</span></li>
<li>10X Genomics Chromium <span class="citation">(Zheng et al. <a href="#ref-Zheng2017aa">2017</a>)</span></li>
</ul>
<p>The methods can be categorized in different ways, but the two most important aspects are <strong>quantification</strong> and <strong>capture</strong>.</p>
<p>For quantification, there are two types, <strong>full-length</strong> and <strong>tag-based</strong>. The former tries to achieve a uniform read coverage of each transcript. By contrast, tag-based protocols only capture either the 5’- or 3’-end of each RNA. The choice of quantification method has important implications for what types of analyses the data can be used for. In theory, full-length protocols should provide an even coverage of transcripts, but as we shall see, there are often biases in the coverage. The main advantage of tag-based protocol is that they can be combined with unique molecular identifiers (UMIs) which can help improve the quantifications. On the other hand, being restricted to one end of the transcript may reduce the mappability and it also makes it harder to distinguish different isoforms <span class="citation">(Archer et al. <a href="#ref-Archer2016-zq">2016</a>)</span>.</p>
<p>The strategy used for capture determines throughput, how the cells can be selected as well as what kind of additional information besides the sequencing that can be obtained. The three most widely used options are <strong>microwell-</strong>, <strong>microfluidic-</strong> and <strong>droplet-</strong> based.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-3"></span>
<img src="figures/300px-Microplates.jpg" alt="Image of microwell plates (image taken from Wikipedia)" width="70%" />
<p class="caption">
Figure 2.4: Image of microwell plates (image taken from Wikipedia)
</p>
</div>
<p>For well-based platforms, cells are isolated using for example pipette or laser capture and placed in microfluidic wells. One advantage of well-based methods is that they can be combined with fluorescent activated cell sorting (FACS), making it possible to select cells based on surface markers. This strategy is thus very useful for situations when one wants to isolate a specific subset of cells for sequencing. Another advantage is that one can take pictures of the cells. The image provides an additional modality and a particularly useful application is to identify wells containg damaged cells or doublets. The main drawback of these methods is that they are often low-throughput and the amount of work required per cell may be considerable.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-4"></span>
<img src="figures/fluidigmC1.jpg" alt="Image of a 96-well Fluidigm C1 chip (image taken from Fluidigm)" width="70%" />
<p class="caption">
Figure 2.5: Image of a 96-well Fluidigm C1 chip (image taken from Fluidigm)
</p>
</div>
<p>Microfluidic platforms, such as <a href="https://www.fluidigm.com/products/c1-system#workflow">Fluidigm’s C1</a>, provide a more integrated system for capturing cells and for carrying out the reactions necessary for the library preparations. Thus, they provide a higher throughput than microwell based platforms. Typically, only around 10% of cells are captured in a microfluidic platform and thus they are not appropriate if one is dealing with rare cell-types or very small amounts of input. Moreover, the chip is relatively expensive, but since reactions can be carried out in a smaller volume money can be saved on reagents.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-5"></span>
<img src="figures/drop-seq.png" alt="Schematic overview of the drop-seq method (Image taken from Macosko et al)" width="60%" />
<p class="caption">
Figure 2.6: Schematic overview of the drop-seq method (Image taken from Macosko et al)
</p>
</div>
<p>The idea behind droplet-based methods is to encapsulate each individual cell inside a nanoliter droplet together with a bead. The bead is loaded with the enzymes required to construct the library. In particular, each bead contains a unique barcode which is attached to all of the reads originating from that cell. Thus, all of the droplets can be pooled, sequenced together and the reads can subsequently be assigned to the cell of origin based on the barcodes. Droplet platforms typically have the highest throughput since the library preparation costs are on the order of <span class="math inline">\(.05\)</span> USD/cell. Instead, sequencing costs often become the limiting factor and a typical experiment the coverage is low with only a few thousand different transcripts detected <span class="citation">(Ziegenhain et al. <a href="#ref-Ziegenhain2017-cu">2017</a>)</span>.</p>
</div>
<div id="what-platform-to-use-for-my-experiment" class="section level2">
<h2><span class="header-section-number">2.7</span> What platform to use for my experiment?</h2>
<p>The most suitable platform depends on the biological question at hand. For example, if one is interested in characterizing the composition of a tissue, then a droplet-based method which will allow a very large number of cells to be captured is likely to be the most appropriate. On the other hand, if one is interesting in characterizing a rare cell-population for which there is a known surface marker, then it is probably best to enrich using FACS and then sequence a smaller number of cells.</p>
<p>Clearly, full-length transcript quantification will be more appropriate if one is interested in studying different isoforms since tagged protocols are much more limited. By contrast, UMIs can only be used with tagged protocols and they can facilitate gene-level quantification.</p>
<p>Two recent studies from the Enard group <span class="citation">(Ziegenhain et al. <a href="#ref-Ziegenhain2017-cu">2017</a>)</span> and the Teichmann group <span class="citation">(Svensson et al. <a href="#ref-Svensson2017-op">2017</a>)</span> have compared several different protocols. In their study, Ziegenhain et al compared five different protocols on the same sample of mouse embryonic stem cells (mESCs). By controlling for the number of cells as well as the sequencing depth, the authors were able to directly compare the sensitivity, noise-levels and costs of the different protocols. One example of their conclusions is illustrated in the figure below which shows the number of genes detected (for a given detection threshold) for the different methods. As you can see, there is almost a two-fold difference between drop-seq and Smart-seq2, suggesting that the choice of protocol can have a major impact on the study</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-6"></span>
<img src="figures/ziegenhainEnardFig1.png" alt="Enard group study" width="60%" />
<p class="caption">
Figure 2.7: Enard group study
</p>
</div>
<p>Svensson et al take a different approach by using synthetic transcripts (spike-ins, more about these later) with known concentrations to measure the accuracy and sensitivity of different protocols. Comparing a wide range of studies, they also reported substantial differences between the protocols.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-7"></span>
<img src="figures/svenssonTeichmannFig2.png" alt="Teichmann group study" width="100%" />
<p class="caption">
Figure 2.8: Teichmann group study
</p>
</div>
<p>As protocols are developed and computational methods for quantifying the technical noise are improved, it is likely that future studies will help us gain further insights regarding the strengths of the different methods. These comparative studies are helpful not only for helping researchers decide which protocol to use, but also for developing new methods as the benchmarking makes it possible to determine what strategies are the most useful ones.</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-Tang2009-bu">
<p>Tang, Fuchou, Catalin Barbacioru, Yangzhou Wang, Ellen Nordman, Clarence Lee, Nanlan Xu, Xiaohui Wang, et al. 2009. “mRNA-Seq Whole-Transcriptome Analysis of a Single Cell.” <em>Nat Meth</em> 6 (5): 377–82. <a href="https://doi.org/10.1038/nmeth.1315">https://doi.org/10.1038/nmeth.1315</a>.</p>
</div>
<div id="ref-Picelli2013-sb">
<p>Picelli, Simone, Åsa K Björklund, Omid R Faridani, Sven Sagasser, Gösta Winberg, and Rickard Sandberg. 2013. “Smart-Seq2 for Sensitive Full-Length Transcriptome Profiling in Single Cells.” <em>Nat Meth</em> 10 (11): 1096–8. <a href="https://doi.org/10.1038/nmeth.2639">https://doi.org/10.1038/nmeth.2639</a>.</p>
</div>
<div id="ref-Hashimshony2012-kd">
<p>Hashimshony, Tamar, Florian Wagner, Noa Sher, and Itai Yanai. 2012. “CEL-Seq: Single-cell RNA-Seq by Multiplexed Linear Amplification.” <em>Cell Reports</em> 2 (3): 666–73. <a href="https://doi.org/10.1016/j.celrep.2012.08.003">https://doi.org/10.1016/j.celrep.2012.08.003</a>.</p>
</div>
<div id="ref-Macosko2015-ix">
<p>Macosko, Evan Z., Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, et al. 2015. “Highly Parallel Genome-Wide Expression Profiling of Individual Cells Using Nanoliter Droplets.” <em>Cell</em> 161 (5): 1202–14. <a href="https://doi.org/10.1016/j.cell.2015.05.002">https://doi.org/10.1016/j.cell.2015.05.002</a>.</p>
</div>
<div id="ref-Stegle2015-uv">
<p>Stegle, Oliver, Sarah A. Teichmann, and John C. Marioni. 2015. “Computational and Analytical Challenges in Single-Cell Transcriptomics.” <em>Nat Rev Genet</em> 16 (3): 133–45. <a href="https://doi.org/10.1038/nrg3833">https://doi.org/10.1038/nrg3833</a>.</p>
</div>
<div id="ref-Kharchenko2014-ts">
<p>Kharchenko, Peter V, Lev Silberstein, and David T Scadden. 2014. “Bayesian Approach to Single-Cell Differential Expression Analysis.” <em>Nat Meth</em> 11 (7): 740–42. <a href="https://doi.org/10.1038/nmeth.2967">https://doi.org/10.1038/nmeth.2967</a>.</p>
</div>
<div id="ref-Hashimshony2016-lx">
<p>Hashimshony, Tamar, Naftalie Senderovich, Gal Avital, Agnes Klochendler, Yaron de Leeuw, Leon Anavy, Dave Gennert, et al. 2016. “CEL-Seq2: Sensitive Highly-Multiplexed Single-Cell RNA-Seq.” <em>Genome Biol</em> 17 (1). <a href="https://doi.org/10.1186/s13059-016-0938-8">https://doi.org/10.1186/s13059-016-0938-8</a>.</p>
</div>
<div id="ref-Klein2015-kz">
<p>Klein, Allon M., Linas Mazutis, Ilke Akartuna, Naren Tallapragada, Adrian Veres, Victor Li, Leonid Peshkin, David A. Weitz, and Marc W. Kirschner. 2015. “Droplet Barcoding for Single-Cell Transcriptomics Applied to Embryonic Stem Cells.” <em>Cell</em> 161 (5): 1187–1201. <a href="https://doi.org/10.1016/j.cell.2015.04.044">https://doi.org/10.1016/j.cell.2015.04.044</a>.</p>
</div>
<div id="ref-Jaitin2014-ko">
<p>Jaitin, Diego Adhemar, Ephraim Kenigsberg, Hadas Keren-Shaul, Naama Elefant, Franziska Paul, Irina Zaretsky, Alexander Mildner, et al. 2014. “Massively Parallel Single-Cell RNA-seq for Marker-Free Decomposition of Tissues into Cell Types.” <em>Science</em> 343 (6172): 776–79.</p>
</div>
<div id="ref-Soumillon2014-eu">
<p>Soumillon, Magali, Davide Cacchiarelli, Stefan Semrau, Alexander van Oudenaarden, and Tarjei S Mikkelsen. 2014. “Characterization of Directed Differentiation by High-Throughput Single-Cell RNA-Seq.” <em>bioRxiv</em>, March, 003236.</p>
</div>
<div id="ref-Gierahn2017-es">
<p>Gierahn, Todd M, Marc H Wadsworth 2nd, Travis K Hughes, Bryan D Bryson, Andrew Butler, Rahul Satija, Sarah Fortune, J Christopher Love, and Alex K Shalek. 2017. “Seq-Well: Portable, Low-Cost RNA Sequencing of Single Cells at High Throughput.” <em>Nat. Methods</em> 14 (4): 395–98.</p>
</div>
<div id="ref-Picelli2014-ic">
<p>Picelli, Simone, Omid R Faridani, Asa K Björklund, Gösta Winberg, Sven Sagasser, and Rickard Sandberg. 2014. “Full-Length RNA-seq from Single Cells Using Smart-Seq2.” <em>Nat. Protoc.</em> 9 (1): 171–81.</p>
</div>
<div id="ref-Islam2014-cn">
<p>Islam, Saiful, Amit Zeisel, Simon Joost, La MannoGioele, Pawel Zajac, Maria Kasper, Peter Lönnerberg, and Sten Linnarsson. 2013. “Quantitative Single-Cell RNA-Seq with Unique Molecular Identifiers.” <em>Nat Meth</em> 11 (2): 163–66. <a href="https://doi.org/10.1038/nmeth.2772">https://doi.org/10.1038/nmeth.2772</a>.</p>
</div>
<div id="ref-Zheng2017aa">
<p>Zheng, Grace X Y, Jessica M Terry, Phillip Belgrader, Paul Ryvkin, Zachary W Bent, Ryan Wilson, Solongo B Ziraldo, et al. 2017. “Massively Parallel Digital Transcriptional Profiling of Single Cells.” <em>Nat Commun</em> 8 (January): 14049. <a href="https://doi.org/10.1038/ncomms14049">https://doi.org/10.1038/ncomms14049</a>.</p>
</div>
<div id="ref-Archer2016-zq">
<p>Archer, Nathan, Mark D. Walsh, Vahid Shahrezaei, and Daniel Hebenstreit. 2016. “Modeling Enzyme Processivity Reveals That RNA-Seq Libraries Are Biased in Characteristic and Correctable Ways.” <em>Cell Systems</em> 3 (5): 467–479.e12. <a href="https://doi.org/10.1016/j.cels.2016.10.012">https://doi.org/10.1016/j.cels.2016.10.012</a>.</p>
</div>
<div id="ref-Ziegenhain2017-cu">
<p>Ziegenhain, Christoph, Beate Vieth, Swati Parekh, Björn Reinius, Amy Guillaumet-Adkins, Martha Smets, Heinrich Leonhardt, Holger Heyn, Ines Hellmann, and Wolfgang Enard. 2017. “Comparative Analysis of Single-Cell RNA Sequencing Methods.” <em>Molecular Cell</em> 65 (4): 631–643.e4. <a href="https://doi.org/10.1016/j.molcel.2017.01.023">https://doi.org/10.1016/j.molcel.2017.01.023</a>.</p>
</div>
<div id="ref-Svensson2017-op">
<p>Svensson, Valentine, Kedar Nath Natarajan, Lam-Ha Ly, Ricardo J Miragaia, Charlotte Labalette, Iain C Macaulay, Ana Cvejic, and Sarah A Teichmann. 2017. “Power Analysis of Single-Cell RNA-Sequencing Experiments.” <em>Nat Meth</em> 14 (4): 381–87. <a href="https://doi.org/10.1038/nmeth.4220">https://doi.org/10.1038/nmeth.4220</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="construction-of-expression-matrix.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["scRNA-seq-course.pdf"],
"toc": {
"collapse": "section"
},
"search": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
