<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop</title>
  <meta name="description" content="Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop" />
  
  
  

<meta name="author" content="Stephanie Hicks (stephaniehicks)">


<meta name="date" content="2018-12-03">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="construction-of-expression-matrix.html">
<link rel="next" href="cleaning-the-expression-matrix.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Welcome!</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#preface"><i class="fa fa-check"></i><b>1.1</b> Preface</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#about-the-course"><i class="fa fa-check"></i><b>1.2</b> About the course</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.3</b> GitHub</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.4</b> Installation</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.6</b> Prerequisites</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.7</b> Contact</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html"><i class="fa fa-check"></i><b>2</b> Introduction to single-cell RNA-seq</a><ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#bulk-rna-seq"><i class="fa fa-check"></i><b>2.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#scrna-seq"><i class="fa fa-check"></i><b>2.2</b> scRNA-seq</a></li>
<li class="chapter" data-level="2.3" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#workflow"><i class="fa fa-check"></i><b>2.3</b> Workflow</a></li>
<li class="chapter" data-level="2.4" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#computational-analysis"><i class="fa fa-check"></i><b>2.4</b> Computational Analysis</a></li>
<li class="chapter" data-level="2.5" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#challenges"><i class="fa fa-check"></i><b>2.5</b> Challenges</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#experimental-methods"><i class="fa fa-check"></i><b>2.6</b> Experimental methods</a></li>
<li class="chapter" data-level="2.7" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#what-platform-to-use-for-my-experiment"><i class="fa fa-check"></i><b>2.7</b> What platform to use for my experiment?</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html"><i class="fa fa-check"></i><b>3</b> Construction of expression matrix</a><ul>
<li class="chapter" data-level="3.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#raw-sequencing-reads-qc"><i class="fa fa-check"></i><b>3.1</b> Raw sequencing reads QC</a><ul>
<li class="chapter" data-level="3.1.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#fastq"><i class="fa fa-check"></i><b>3.1.1</b> FastQ</a></li>
<li class="chapter" data-level="3.1.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#check-the-quality-of-the-raw-sequencing-reads"><i class="fa fa-check"></i><b>3.1.2</b> Check the quality of the raw sequencing reads</a></li>
<li class="chapter" data-level="3.1.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#trim-sequencing-adapaters-from-sequencing-reads"><i class="fa fa-check"></i><b>3.1.3</b> Trim sequencing adapaters from sequencing reads</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-alignment"><i class="fa fa-check"></i><b>3.2</b> Reads alignment</a><ul>
<li class="chapter" data-level="3.2.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#genome-fasta-gtf"><i class="fa fa-check"></i><b>3.2.1</b> Genome (<code>FASTA</code>, <code>GTF</code>)</a></li>
<li class="chapter" data-level="3.2.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#alignment-using-star"><i class="fa fa-check"></i><b>3.2.2</b> Alignment using Star</a></li>
<li class="chapter" data-level="3.2.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#pseudo-alignment-using-salmon"><i class="fa fa-check"></i><b>3.2.3</b> Pseudo-alignment using Salmon</a></li>
<li class="chapter" data-level="3.2.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapped-reads-bam-file-format"><i class="fa fa-check"></i><b>3.2.4</b> Mapped reads: <code>BAM</code> file format</a></li>
<li class="chapter" data-level="3.2.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cram"><i class="fa fa-check"></i><b>3.2.5</b> CRAM</a></li>
<li class="chapter" data-level="3.2.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-mapping-qc"><i class="fa fa-check"></i><b>3.2.6</b> Reads Mapping QC</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-quantification"><i class="fa fa-check"></i><b>3.3</b> Reads quantification</a></li>
<li class="chapter" data-level="3.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#unique-molecular-identifiers-umis"><i class="fa fa-check"></i><b>3.4</b> Unique Molecular Identifiers (UMIs)</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html"><i class="fa fa-check"></i><b>4</b> Introduction to R/Bioconductor</a><ul>
<li class="chapter" data-level="4.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#installing-r-packages"><i class="fa fa-check"></i><b>4.1</b> Installing R packages</a><ul>
<li class="chapter" data-level="4.1.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#cran"><i class="fa fa-check"></i><b>4.1.1</b> CRAN</a></li>
<li class="chapter" data-level="4.1.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#github-1"><i class="fa fa-check"></i><b>4.1.2</b> Github</a></li>
<li class="chapter" data-level="4.1.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#bioconductor"><i class="fa fa-check"></i><b>4.1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="4.1.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#source"><i class="fa fa-check"></i><b>4.1.4</b> Source</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#data-types"><i class="fa fa-check"></i><b>4.2</b> Data Types</a><ul>
<li class="chapter" data-level="4.2.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#what-is-tidy-data"><i class="fa fa-check"></i><b>4.2.1</b> What is Tidy Data?</a></li>
<li class="chapter" data-level="4.2.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#what-is-rich-data"><i class="fa fa-check"></i><b>4.2.2</b> What is Rich Data?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#bioconductor-1"><i class="fa fa-check"></i><b>4.3</b> Bioconductor</a><ul>
<li class="chapter" data-level="4.3.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#singlecellexperiment-class"><i class="fa fa-check"></i><b>4.3.1</b> <code>SingleCellExperiment</code> class</a></li>
<li class="chapter" data-level="4.3.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#scater-package"><i class="fa fa-check"></i><b>4.3.2</b> <code>scater</code> package</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#data-visualization-using-ggplot2"><i class="fa fa-check"></i><b>4.4</b> Data visualization using ggplot2</a><ul>
<li class="chapter" data-level="4.4.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#what-is-ggplot2"><i class="fa fa-check"></i><b>4.4.1</b> What is ggplot2?</a></li>
<li class="chapter" data-level="4.4.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#principles-of-ggplot2"><i class="fa fa-check"></i><b>4.4.2</b> Principles of ggplot2</a></li>
<li class="chapter" data-level="4.4.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#using-the-aes-mapping-function"><i class="fa fa-check"></i><b>4.4.3</b> Using the <code>aes()</code> mapping function</a></li>
<li class="chapter" data-level="4.4.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#geoms"><i class="fa fa-check"></i><b>4.4.4</b> Geoms</a></li>
<li class="chapter" data-level="4.4.5" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#plotting-data-from-more-than-2-cells"><i class="fa fa-check"></i><b>4.4.5</b> Plotting data from more than 2 cells</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html"><i class="fa fa-check"></i><b>5</b> Cleaning the expression matrix</a><ul>
<li class="chapter" data-level="5.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#expression-qc"><i class="fa fa-check"></i><b>5.1</b> Expression QC</a><ul>
<li class="chapter" data-level="5.1.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction"><i class="fa fa-check"></i><b>5.1.1</b> Introduction</a></li>
<li class="chapter" data-level="5.1.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#tung-dataset-umi"><i class="fa fa-check"></i><b>5.1.2</b> Tung dataset (UMI)</a></li>
<li class="chapter" data-level="5.1.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cell-qc"><i class="fa fa-check"></i><b>5.1.3</b> Cell QC</a></li>
<li class="chapter" data-level="5.1.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cell-filtering"><i class="fa fa-check"></i><b>5.1.4</b> Cell filtering</a></li>
<li class="chapter" data-level="5.1.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#compare-filterings"><i class="fa fa-check"></i><b>5.1.5</b> Compare filterings</a></li>
<li class="chapter" data-level="5.1.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#gene-analysis"><i class="fa fa-check"></i><b>5.1.6</b> Gene analysis</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#data-visualization"><i class="fa fa-check"></i><b>5.2</b> Data visualization</a><ul>
<li class="chapter" data-level="5.2.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-1"><i class="fa fa-check"></i><b>5.2.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#pca-plot"><i class="fa fa-check"></i><b>5.2.2</b> PCA plot</a></li>
<li class="chapter" data-level="5.2.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#tsne-map"><i class="fa fa-check"></i><b>5.2.3</b> tSNE map</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#identifying-confounding-factors"><i class="fa fa-check"></i><b>5.3</b> Identifying confounding factors</a><ul>
<li class="chapter" data-level="5.3.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-2"><i class="fa fa-check"></i><b>5.3.1</b> Introduction</a></li>
<li class="chapter" data-level="5.3.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#correlations-with-pcs"><i class="fa fa-check"></i><b>5.3.2</b> Correlations with PCs</a></li>
<li class="chapter" data-level="5.3.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#explanatory-variables"><i class="fa fa-check"></i><b>5.3.3</b> Explanatory variables</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalization-theory"><i class="fa fa-check"></i><b>5.4</b> Normalization theory</a><ul>
<li class="chapter" data-level="5.4.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-3"><i class="fa fa-check"></i><b>5.4.1</b> Introduction</a></li>
<li class="chapter" data-level="5.4.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#library-size-1"><i class="fa fa-check"></i><b>5.4.2</b> Library size</a></li>
<li class="chapter" data-level="5.4.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalisations"><i class="fa fa-check"></i><b>5.4.3</b> Normalisations</a></li>
<li class="chapter" data-level="5.4.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#effectiveness"><i class="fa fa-check"></i><b>5.4.4</b> Effectiveness</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#normalization-practice"><i class="fa fa-check"></i><b>5.5</b> Normalization practice</a><ul>
<li class="chapter" data-level="5.5.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#raw"><i class="fa fa-check"></i><b>5.5.1</b> Raw</a></li>
<li class="chapter" data-level="5.5.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#cpm-1"><i class="fa fa-check"></i><b>5.5.2</b> CPM</a></li>
<li class="chapter" data-level="5.5.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#scran-1"><i class="fa fa-check"></i><b>5.5.3</b> scran</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#dealing-with-confounders"><i class="fa fa-check"></i><b>5.6</b> Dealing with confounders</a><ul>
<li class="chapter" data-level="5.6.1" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#introduction-4"><i class="fa fa-check"></i><b>5.6.1</b> Introduction</a></li>
<li class="chapter" data-level="5.6.2" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#combat"><i class="fa fa-check"></i><b>5.6.2</b> Combat</a></li>
<li class="chapter" data-level="5.6.3" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#mnncorrect"><i class="fa fa-check"></i><b>5.6.3</b> mnnCorrect</a></li>
<li class="chapter" data-level="5.6.4" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#glm"><i class="fa fa-check"></i><b>5.6.4</b> GLM</a></li>
<li class="chapter" data-level="5.6.5" data-path="cleaning-the-expression-matrix.html"><a href="cleaning-the-expression-matrix.html#how-to-evaluate-and-compare-confounder-removal-strategies"><i class="fa fa-check"></i><b>5.6.5</b> How to evaluate and compare confounder removal strategies</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="biological-analysis.html"><a href="biological-analysis.html"><i class="fa fa-check"></i><b>6</b> Biological Analysis</a><ul>
<li class="chapter" data-level="6.1" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering"><i class="fa fa-check"></i><b>6.1</b> Clustering</a><ul>
<li class="chapter" data-level="6.1.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-5"><i class="fa fa-check"></i><b>6.1.1</b> Introduction</a></li>
<li class="chapter" data-level="6.1.2" data-path="biological-analysis.html"><a href="biological-analysis.html#dimensionality-reduction"><i class="fa fa-check"></i><b>6.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.1.3" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-methods"><i class="fa fa-check"></i><b>6.1.3</b> Clustering methods</a></li>
<li class="chapter" data-level="6.1.4" data-path="biological-analysis.html"><a href="biological-analysis.html#challenges-in-clustering"><i class="fa fa-check"></i><b>6.1.4</b> Challenges in clustering</a></li>
<li class="chapter" data-level="6.1.5" data-path="biological-analysis.html"><a href="biological-analysis.html#software-packages-for-scrna-seq-data"><i class="fa fa-check"></i><b>6.1.5</b> Software packages for scRNA-seq data</a></li>
<li class="chapter" data-level="6.1.6" data-path="biological-analysis.html"><a href="biological-analysis.html#comparing-clustering"><i class="fa fa-check"></i><b>6.1.6</b> Comparing clustering</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-example"><i class="fa fa-check"></i><b>6.2</b> Clustering example</a><ul>
<li class="chapter" data-level="6.2.1" data-path="biological-analysis.html"><a href="biological-analysis.html#deng-dataset"><i class="fa fa-check"></i><b>6.2.1</b> Deng dataset</a></li>
<li class="chapter" data-level="6.2.2" data-path="biological-analysis.html"><a href="biological-analysis.html#sc3-1"><i class="fa fa-check"></i><b>6.2.2</b> SC3</a></li>
<li class="chapter" data-level="6.2.3" data-path="biological-analysis.html"><a href="biological-analysis.html#tsne-kmeans"><i class="fa fa-check"></i><b>6.2.3</b> tSNE + kmeans</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="biological-analysis.html"><a href="biological-analysis.html#feature-selection"><i class="fa fa-check"></i><b>6.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="6.3.1" data-path="biological-analysis.html"><a href="biological-analysis.html#identifying-genes-vs-a-null-model"><i class="fa fa-check"></i><b>6.3.1</b> Identifying Genes vs a Null Model</a></li>
<li class="chapter" data-level="6.3.2" data-path="biological-analysis.html"><a href="biological-analysis.html#correlated-expression"><i class="fa fa-check"></i><b>6.3.2</b> Correlated Expression</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="biological-analysis.html"><a href="biological-analysis.html#differential-expression-de-analysis"><i class="fa fa-check"></i><b>6.4</b> Differential Expression (DE) analysis</a><ul>
<li class="chapter" data-level="6.4.1" data-path="biological-analysis.html"><a href="biological-analysis.html#bulk-rna-seq-1"><i class="fa fa-check"></i><b>6.4.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="6.4.2" data-path="biological-analysis.html"><a href="biological-analysis.html#single-cell-rna-seq"><i class="fa fa-check"></i><b>6.4.2</b> Single-cell RNA-seq</a></li>
<li class="chapter" data-level="6.4.3" data-path="biological-analysis.html"><a href="biological-analysis.html#differences-in-distribution"><i class="fa fa-check"></i><b>6.4.3</b> Differences in distribution</a></li>
<li class="chapter" data-level="6.4.4" data-path="biological-analysis.html"><a href="biological-analysis.html#models-of-single-cell-rna-seq-data"><i class="fa fa-check"></i><b>6.4.4</b> Models of single-cell RNA-seq data</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="biological-analysis.html"><a href="biological-analysis.html#de-in-tung-dataset"><i class="fa fa-check"></i><b>6.5</b> DE in <code>tung</code> dataset</a><ul>
<li class="chapter" data-level="6.5.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-6"><i class="fa fa-check"></i><b>6.5.1</b> Introduction</a></li>
<li class="chapter" data-level="6.5.2" data-path="biological-analysis.html"><a href="biological-analysis.html#kolmogorov-smirnov-test"><i class="fa fa-check"></i><b>6.5.2</b> Kolmogorov-Smirnov test</a></li>
<li class="chapter" data-level="6.5.3" data-path="biological-analysis.html"><a href="biological-analysis.html#wilcoxmann-whitney-u-test"><i class="fa fa-check"></i><b>6.5.3</b> Wilcox/Mann-Whitney-U Test</a></li>
<li class="chapter" data-level="6.5.4" data-path="biological-analysis.html"><a href="biological-analysis.html#edger-or-deseq2"><i class="fa fa-check"></i><b>6.5.4</b> edgeR or DESeq2</a></li>
<li class="chapter" data-level="6.5.5" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle"><i class="fa fa-check"></i><b>6.5.5</b> Monocle</a></li>
<li class="chapter" data-level="6.5.6" data-path="biological-analysis.html"><a href="biological-analysis.html#mast"><i class="fa fa-check"></i><b>6.5.6</b> MAST</a></li>
<li class="chapter" data-level="6.5.7" data-path="biological-analysis.html"><a href="biological-analysis.html#scde"><i class="fa fa-check"></i><b>6.5.7</b> SCDE</a></li>
<li class="chapter" data-level="6.5.8" data-path="biological-analysis.html"><a href="biological-analysis.html#zinbwave-deseq2edger"><i class="fa fa-check"></i><b>6.5.8</b> zinbwave + DESeq2/edger</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="http://www.sanger.ac.uk/science/groups/hemberg-group" target="blank">Hemberg Lab</a></li>
<li><a href="http://www.stephaniehicks.comp" target="blank">Stephanie Hicks</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis of single cell RNA-seq data: 2018 BioInfoSummer Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-rbioconductor" class="section level1">
<h1><span class="header-section-number">4</span> Introduction to R/Bioconductor</h1>
<div id="installing-r-packages" class="section level2">
<h2><span class="header-section-number">4.1</span> Installing R packages</h2>
<div id="cran" class="section level3">
<h3><span class="header-section-number">4.1.1</span> CRAN</h3>
<p>The Comprehensive R Archive Network <a href="https://cran.r-project.org/">CRAN</a>
is the biggest archive of R packages. There are few requirements for
uploading packages besides building and installing succesfully, hence
documentation and support is often minimal and figuring how to use
these packages can be a challenge it itself. CRAN is the default
repository R will search to find packages to install:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="kw">install.packages</span>(<span class="st">&quot;RColorBrewer&quot;</span>, <span class="st">&quot;reshape2&quot;</span>, </a>
<a class="sourceLine" id="cb6-2" title="2">              <span class="st">&quot;matrixStats&quot;</span>, <span class="st">&quot;mclust&quot;</span>, <span class="st">&quot;pheatmap&quot;</span>, <span class="st">&quot;mvoutlier&quot;</span>)</a>
<a class="sourceLine" id="cb6-3" title="3"><span class="kw">library</span>(matrixStats)</a></code></pre></div>
</div>
<div id="github-1" class="section level3">
<h3><span class="header-section-number">4.1.2</span> Github</h3>
<p><a href="https://github.com/">Github</a> isn’t specific to R, any code of any
type in any state can be uploaded. There is no guarantee a package
uploaded to github will even install, nevermind do what it claims
to do. R packages can be downloaded and installed directly from
github using the <code>devtools</code> package installed above.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</a>
<a class="sourceLine" id="cb7-2" title="2"><span class="kw">library</span>(<span class="st">&quot;devtools&quot;</span>)</a>
<a class="sourceLine" id="cb7-3" title="3"></a>
<a class="sourceLine" id="cb7-4" title="4">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;hemberg-lab/scRNA.seq.funcs&quot;</span>)</a>
<a class="sourceLine" id="cb7-5" title="5">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;theislab/kBET&quot;</span>)</a></code></pre></div>
<p>Github is also a version control system which stores multiple
versions of any package. By default the most recent “master”
version of the package is installed. If you want an older
version or the development branch this can be specified
using the “ref” parameter:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co"># different branch</span></a>
<a class="sourceLine" id="cb8-2" title="2">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;path/to/github/repo&quot;</span>, <span class="dt">ref=</span><span class="st">&quot;older-branch&quot;</span>)</a>
<a class="sourceLine" id="cb8-3" title="3"><span class="co"># previous commit</span></a>
<a class="sourceLine" id="cb8-4" title="4">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;path/to/github/repo&quot;</span>, <span class="dt">ref=</span><span class="st">&quot;previous-commit-id&quot;</span>)</a></code></pre></div>
</div>
<div id="bioconductor" class="section level3">
<h3><span class="header-section-number">4.1.3</span> Bioconductor</h3>
<p>Bioconductor is an open-source, open-development
repository of R-packages specifically for high-throughput
biological analyses. It has the strictest requirements for
submission, including installation on every platform and
full documentation with a tutorial (called a vignette)
explaining how the package should be used. Bioconductor
also encourages utilization of standard data structures/classes
and coding style/naming conventions, so that, in theory,
packages and analyses can be combined into large pipelines
or workflows.</p>
<p>This is the old way of installing packages:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)</a>
<a class="sourceLine" id="cb9-2" title="2"><span class="kw">biocLite</span>(<span class="st">&quot;edgeR&quot;</span>)</a></code></pre></div>
<p>Note: in some situations it is necessary to substitute
“<a href="http://" class="uri">http://</a>” for “<a href="https://" class="uri">https://</a>” in the above depending on the
security features of your internet connection/network.</p>
<p>However, this is the new (and improved way) of installing
Bioconductor packages using <a href="https://cran.r-project.org/web/packages/BiocManager/index.html">BiocManager</a>:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</a>
<a class="sourceLine" id="cb10-2" title="2"><span class="kw">library</span>(<span class="st">&quot;BiocManager&quot;</span>)</a>
<a class="sourceLine" id="cb10-3" title="3"></a>
<a class="sourceLine" id="cb10-4" title="4">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;scran&quot;</span>, <span class="st">&quot;Rtsne&quot;</span>, <span class="st">&quot;sva&quot;</span>, <span class="st">&quot;DESeq2&quot;</span>, <span class="st">&quot;edgeR&quot;</span>, <span class="st">&quot;SC3&quot;</span>, <span class="st">&quot;zinbwave&quot;</span>)</a></code></pre></div>
<p>You can even install packages from a specific version of Bioconductor:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;scater&quot;</span>, <span class="dt">version=</span><span class="st">&quot;devel&quot;</span>) <span class="co"># development branch</span></a>
<a class="sourceLine" id="cb11-2" title="2">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;scater&quot;</span>, <span class="dt">version=</span><span class="st">&quot;3.8&quot;</span>) <span class="co"># current release as of Fall 2018</span></a></code></pre></div>
<p>Bioconductor also requires creators to support their
packages and has a regular 6-month release schedule.
Make sure you are using the most recent release of
bioconductor before trying to install packages for the course.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">BiocManager<span class="op">::</span><span class="kw">install</span>()</a></code></pre></div>
</div>
<div id="source" class="section level3">
<h3><span class="header-section-number">4.1.4</span> Source</h3>
<p>The final way to install packages is directly from source.
In this case you have to download a fully built source
code file, usually packagename.tar.gz, or clone the github
repository and rebuild the package yourself. Generally this
will only be done if you want to edit a package yourself,
or if for some reason the former methods have failed.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">pkgs &lt;-<span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>())</a>
<a class="sourceLine" id="cb13-2" title="2">BiocManager<span class="op">::</span><span class="kw">install</span>(pkgs, <span class="dt">type =</span> <span class="st">&quot;source&quot;</span>)</a></code></pre></div>
</div>
</div>
<div id="data-types" class="section level2">
<h2><span class="header-section-number">4.2</span> Data Types</h2>
<div id="what-is-tidy-data" class="section level3">
<h3><span class="header-section-number">4.2.1</span> What is Tidy Data?</h3>
<p>The <a href="https://www.tidyverse.org">tidyverse</a> is <em>“an opinionated
collection of R packages designed for data science. All packages
share an underlying philosophy and common APIs.”</em></p>
<p>Another way of putting it is that it’s a set of packages
that are useful specifically for data manipulation,
exploration and visualization with a common philosphy.</p>
<div id="what-is-this-common-philosphy" class="section level4">
<h4><span class="header-section-number">4.2.1.1</span> What is this common philosphy?</h4>
<p>The common philosphy is called <em>“tidy”</em> data. It is
a standard way of mapping the meaning of a dataset
to its structure.</p>
<p>In <em>tidy</em> data:</p>
<ul>
<li>Each variable forms a column.</li>
<li>Each observation forms a row.</li>
<li>Each type of observational unit forms a table.</li>
</ul>
<p><img src="http://r4ds.had.co.nz/images/tidy-1.png" width="95%" style="display: block; margin: auto;" /></p>
<p>Below, we are interested in transformating the table on
the right to the the table on the left, which is
considered “tidy”.</p>
<p><img src="http://r4ds.had.co.nz/images/tidy-9.png" width="95%" style="display: block; margin: auto;" /></p>
<p>Working with tidy data is useful because it creates a structured way of
organizing data values within a data set. This makes the data analysis
process more efficient and simplifies the development of data analysis tools
that work together. In this way, you can focus on the problem you are
investigating, rather than the uninteresting logistics of data.</p>
</div>
<div id="what-is-in-the-tidyverse" class="section level4">
<h4><span class="header-section-number">4.2.1.2</span> What is in the <code>tidyverse</code>?</h4>
<p>We can install and load the set of R packages using
<code>install.packages(&quot;tidyverse&quot;)</code> function.</p>
<p>When we load the tidyverse package using <code>library(tidyverse)</code>,
there are six core R packages that load:</p>
<ul>
<li><a href="http://readr.tidyverse.org">readr</a>, for data import.</li>
<li><a href="http://tidyr.tidyverse.org">tidyr</a>, for data tidying.</li>
<li><a href="http://dplyr.tidyverse.org">dplyr</a>, for data wrangling.</li>
<li><a href="http://ggplot2.tidyverse.org">ggplot2</a>, for data visualisation.</li>
<li><a href="http://purrr.tidyverse.org">purrr</a>, for functional programming.</li>
<li><a href="http://tibble.tidyverse.org">tibble</a>, for tibbles, a modern re-imagining of data frames.</li>
</ul>
<p>Here, we load in the tidyverse.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="kw">library</span>(tidyverse)</a></code></pre></div>
<p>These packages are highlighted in bold here:</p>
<p><img src="https://rviews.rstudio.com/post/2017-06-09-What-is-the-tidyverse_files/tidyverse1.png" width="95%" style="display: block; margin: auto;" /></p>
<p>Because these packages all share the “tidy” philosphy,
the data analysis workflow is easier as you move from
package to package.</p>
<p>If we are
going to take advantage of the <em>“tidyverse”</em>,
this means we need to <em>transform</em> the data
into a form that is <em>“tidy”</em>. If you recall,
in <em>tidy</em> data:</p>
<ul>
<li>Each variable forms a column.</li>
<li>Each observation forms a row.</li>
<li>Each type of observational unit forms a table.</li>
</ul>
<p>For example, consider the following dataset:</p>
<p><img src="https://github.com/datasciencelabs/2016/raw/master/lectures/wrangling/pics/stocks-by-company.png" /></p>
<p>Here:</p>
<ul>
<li>each row represents one company (row names are companies)</li>
<li>each column represent one time point</li>
<li>the stock prices are defined for each row/column pair</li>
</ul>
<p>Alternatively, a data set can be structured in the following way:</p>
<ul>
<li>each row represents one time point (but no row names)</li>
<li>the first column defines the time variable and the last three columns contain the stock prices for three companies</li>
</ul>
<p><img src="https://github.com/datasciencelabs/2016/raw/master/lectures/wrangling/pics/stocks-by-time.png" /></p>
<p>In both cases, the data is the same, but the structure is
different. This can be <em>frustrating</em> to deal with as an
analyst because the meaning of the values (rows and columns)
in the two data sets are different. Providing a standardized
way of organizing values within a data set would alleviate
a major portion of this frustration.</p>
<p>For motivation, a <em>tidy</em> version of the stock data we
looked at above looks like this: (we’ll learn how the
functions work in just a moment)</p>
<p><img src="https://github.com/datasciencelabs/2016/raw/master/lectures/wrangling/pics/stocks-tidy.png" /></p>
<p>In this “tidy” data set, we have three columns representing
three variables (time, company name and stock price).
Every row represents contains one stock price from a
particular time and for a specific company.</p>
<p>The <a href="https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html"><code>tidyr</code></a>
is an R package that transforms data sets to a tidy format.</p>
</div>
</div>
<div id="what-is-rich-data" class="section level3">
<h3><span class="header-section-number">4.2.2</span> What is Rich Data?</h3>
<p>If you google ‘rich data’, you will find lots of different definitions
for this term. In this course, we will use ‘rich data’ to mean data which
is generated by combining information from multiple sources. For example,
you could make rich data by creating an object in R which contains a matrix
of gene expression values across the cells in your scRNA-seq
experiment, but also information about how the experiment was performed.
Objects of the <code>SingleCellExperiment</code> class, which we will discuss below,
are an example of rich data.</p>
</div>
</div>
<div id="bioconductor-1" class="section level2">
<h2><span class="header-section-number">4.3</span> Bioconductor</h2>
<p>From <a href="https://en.wikipedia.org/wiki/Bioconductor">Wikipedia</a>:
<a href="https://www.bioconductor.org/">Bioconductor</a> is a free, open source and open
development software project for the analysis and comprehension of genomic data
generated by wet lab experiments in molecular biology. Bioconductor is based
primarily on the statistical R programming language, but does contain
contributions in other programming languages. It has two releases each year that
follow the semiannual releases of R. At any one time there is a release version,
which corresponds to the released version of R, and a development version, which
corresponds to the development version of R. Most users will find the release
version appropriate for their needs.</p>
<p>We strongly recommend all new comers and even experienced high-throughput data
analysts to use well developed and maintained
<a href="https://www.bioconductor.org/developers/how-to/commonMethodsAndClasses/">Bioconductor methods and classes</a>.</p>
<div id="singlecellexperiment-class" class="section level3">
<h3><span class="header-section-number">4.3.1</span> <code>SingleCellExperiment</code> class</h3>
<p><a href="http://bioconductor.org/packages/SingleCellExperiment"><code>SingleCellExperiment</code></a>
(SCE) is a S4 class for storing data from single-cell experiments. This includes
specialized methods to store and retrieve spike-in information, dimensionality
reduction coordinates and size factors for each cell, along with the usual
metadata for genes and libraries.</p>
<p><img src="figures/SCE.jpg" width="95%" style="display: block; margin: auto;" /></p>
<p>In practice, an object of this class can be created using the
<code>SingleCellExperiment()</code> constructor:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="kw">library</span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb15-2" title="2">counts &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rpois</span>(<span class="dv">100</span>, <span class="dt">lambda =</span> <span class="dv">10</span>), <span class="dt">ncol=</span><span class="dv">10</span>, <span class="dt">nrow=</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb15-3" title="3"><span class="kw">rownames</span>(counts) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb15-4" title="4"><span class="kw">colnames</span>(counts) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;cell&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb15-5" title="5">sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(</a>
<a class="sourceLine" id="cb15-6" title="6">    <span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> counts),</a>
<a class="sourceLine" id="cb15-7" title="7">    <span class="dt">rowData =</span> <span class="kw">data.frame</span>(<span class="dt">gene_names =</span> <span class="kw">paste</span>(<span class="st">&quot;gene_name&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)),</a>
<a class="sourceLine" id="cb15-8" title="8">    <span class="dt">colData =</span> <span class="kw">data.frame</span>(<span class="dt">cell_names =</span> <span class="kw">paste</span>(<span class="st">&quot;cell_name&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>))</a>
<a class="sourceLine" id="cb15-9" title="9">)</a>
<a class="sourceLine" id="cb15-10" title="10">sce</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 10 10 
## metadata(0):
## assays(1): counts
## rownames(10): gene1 gene2 ... gene9 gene10
## rowData names(1): gene_names
## colnames(10): cell1 cell2 ... cell9 cell10
## colData names(1): cell_names
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<p>In the <code>SingleCellExperiment</code> object, users can assign arbitrary names to
entries of assays. To assist interoperability between packages, some
suggestions for what the names should be for particular types of data
are provided by the authors:</p>
<ul>
<li><strong>counts</strong>: Raw count data, e.g., number of reads or transcripts for a particular gene.</li>
<li><strong>normcounts</strong>: Normalized values on the same scale as the original counts. For example, counts divided by cell-specific size factors that are centred at unity.</li>
<li><strong>logcounts</strong>: Log-transformed counts or count-like values. In most cases, this will be defined as log-transformed normcounts, e.g., using log base 2 and a pseudo-count of 1.</li>
<li><strong>cpm</strong>: Counts-per-million. This is the read count for each gene in each cell, divided by the library size of each cell in millions.</li>
<li><strong>tpm</strong>: Transcripts-per-million. This is the number of transcripts for each gene in each cell, divided by the total number of transcripts in that cell (in millions).</li>
</ul>
<p>Each of these suggested names has an appropriate getter/setter method
for convenient manipulation of the <code>SingleCellExperiment</code>. For example,
we can take the (very specifically named) <code>counts</code> slot, normalize it and
assign it to <code>normcounts</code> instead:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1"><span class="kw">normcounts</span>(sce) &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="kw">counts</span>(sce) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb17-2" title="2">sce</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 10 10 
## metadata(0):
## assays(2): counts normcounts
## rownames(10): gene1 gene2 ... gene9 gene10
## rowData names(1): gene_names
## colnames(10): cell1 cell2 ... cell9 cell10
## colData names(1): cell_names
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1"><span class="kw">dim</span>(<span class="kw">normcounts</span>(sce))</a></code></pre></div>
<pre><code>## [1] 10 10</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1"><span class="kw">head</span>(<span class="kw">normcounts</span>(sce))</a></code></pre></div>
<pre><code>##          cell1    cell2    cell3    cell4    cell5    cell6    cell7
## gene1 3.169925 3.321928 3.321928 3.321928 4.000000 3.169925 3.321928
## gene2 3.169925 3.169925 3.169925 3.321928 4.000000 3.459432 3.459432
## gene3 3.807355 3.169925 3.906891 4.247928 3.000000 3.000000 3.459432
## gene4 3.584963 3.321928 3.906891 3.807355 3.169925 3.321928 3.807355
## gene5 3.906891 3.459432 3.321928 3.169925 4.169925 3.169925 3.584963
## gene6 4.169925 3.000000 3.000000 3.169925 4.169925 3.169925 3.700440
##          cell8    cell9   cell10
## gene1 3.700440 3.700440 3.321928
## gene2 3.906891 3.700440 3.906891
## gene3 3.584963 3.459432 3.000000
## gene4 4.087463 3.000000 3.169925
## gene5 4.087463 3.584963 3.321928
## gene6 3.169925 2.584963 3.700440</code></pre>
</div>
<div id="scater-package" class="section level3">
<h3><span class="header-section-number">4.3.2</span> <code>scater</code> package</h3>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1"><span class="kw">library</span>(scater)</a></code></pre></div>
<p><a href="http://bioconductor.org/packages/scater/"><code>scater</code></a> is a R package for single-cell RNA-seq analysis <span class="citation">(McCarthy et al. <a href="#ref-McCarthy2017-kb">2017</a>)</span>. The package contains several useful methods for quality control, visualisation and pre-processing of data prior to further downstream analysis.</p>
<p><code>scater</code> features the following functionality:</p>
<ul>
<li>Automated computation of QC metrics</li>
<li>Transcript quantification from read data with pseudo-alignment</li>
<li>Data format standardisation</li>
<li>Rich visualizations for exploratory analysis</li>
<li>Seamless integration into the Bioconductor universe</li>
<li>Simple normalisation methods</li>
</ul>
<p>We highly recommend to use <code>scater</code> for all single-cell RNA-seq analyses and
<code>scater</code> is the basis of the first part of the course.</p>
<p>As illustrated in the figure below, <code>scater</code> will help you with quality control,
filtering and normalization of your expression matrix following mapping
and alignment. <span style="color:red">Keep in mind that this figure
represents the original version of <code>scater</code> where an <code>SCESet</code> class was used.
In the newest version this figure is still correct, except that <code>SCESet</code>
can be substituted with the <code>SingleCellExperiment</code> class.</span></p>
<p><img src="figures/scater_qc_workflow.png" width="90%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="data-visualization-using-ggplot2" class="section level2">
<h2><span class="header-section-number">4.4</span> Data visualization using ggplot2</h2>
<div id="what-is-ggplot2" class="section level3">
<h3><span class="header-section-number">4.4.1</span> What is ggplot2?</h3>
<p><code>ggplot2</code> is an R package designed by Hadley Wickham which facilitates
data visualizations. In this section, we will touch briefly on some of the
features of the package. If you would like to learn more about how to use
ggplot2, we would recommend reading “ggplot2 Elegant graphics for data analysis”,
by Hadley Wickham.</p>
</div>
<div id="principles-of-ggplot2" class="section level3">
<h3><span class="header-section-number">4.4.2</span> Principles of ggplot2</h3>
<ul>
<li>Your data must be a <code>data.frame</code> if you want to plot it using <code>ggplot2</code>.</li>
<li>Use the <code>aes</code> mapping function to specify how variables in the <code>data.frame</code> map to features on your plot</li>
<li>Use <code>geoms</code> to specify how your data should be represented on your graph eg. as a scatterplot, a barplot, a boxplot etc.</li>
</ul>
</div>
<div id="using-the-aes-mapping-function" class="section level3">
<h3><span class="header-section-number">4.4.3</span> Using the <code>aes()</code> mapping function</h3>
<p>The <code>aes</code> function specifies how variables in your dataframe map to
features on your plot. To understand how this works, let’s look at an example:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" title="1"><span class="kw">library</span>(tidyverse) <span class="co"># this loads the ggplot2 R package</span></a>
<a class="sourceLine" id="cb24-2" title="2"><span class="kw">set.seed</span>(<span class="dv">12345</span>)</a>
<a class="sourceLine" id="cb24-3" title="3"></a>
<a class="sourceLine" id="cb24-4" title="4">counts &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">matrix</span>(<span class="kw">rpois</span>(<span class="dv">100</span>, <span class="dt">lambda =</span> <span class="dv">10</span>), <span class="dt">ncol=</span><span class="dv">10</span>, <span class="dt">nrow=</span><span class="dv">10</span>))</a>
<a class="sourceLine" id="cb24-5" title="5">gene_ids &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb24-6" title="6"><span class="kw">colnames</span>(counts) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;cell&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb24-7" title="7">counts &lt;-<span class="st"> </span><span class="kw">data.frame</span>(gene_ids, counts)</a>
<a class="sourceLine" id="cb24-8" title="8">counts</a></code></pre></div>
<pre><code>##    gene_ids cell1 cell2 cell3 cell4 cell5 cell6 cell7 cell8 cell9 cell10
## 1     gene1    11     9    15    11    12    16     9     7    13     14
## 2     gene2    12    10     8     8     9    12     8    15     8      8
## 3     gene3     9     9    11     4    10     7    10    12     2     11
## 4     gene4     8    11    11    11    10     4     8     9     7      6
## 5     gene5    11    13     9     6    14    11    12    11    13     10
## 6     gene6     4    10    12    18    13    13     9    12    10     11
## 7     gene7    11    12    16    10     7     7    10     6    12      9
## 8     gene8     7    14    16     5    12    10    10     7     5     10
## 9     gene9     8     7    15     5    12     7    12     8     8     10
## 10   gene10    11     4    10    10    16    15     9     6     9     11</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> cell1, <span class="dt">y =</span> cell2))</a></code></pre></div>
<p><img src="04-r-and-bioconductor_files/figure-html/unnamed-chunk-19-1.png" width="90%" style="display: block; margin: auto;" /></p>
<p>Let’s take a closer look at the final command,
<code>ggplot(data = counts, mapping = aes(x = cell1, y = cell2))</code>. <code>ggplot()</code>
initializes a ggplot object and takes the arguments <code>data</code> and <code>mapping</code>.
We pass our dataframe of counts to <code>data</code> and use the <code>aes()</code> function
to specify that we would like to use the variable <code>cell1</code> as our <code>x</code>
variable and the variable <code>cell2</code> as our <code>y</code> variable.</p>
<p>Clearly, the plot we have just created is not very informative because
no data is displayed on them. To display data, we will need to use <code>geoms</code>.</p>
</div>
<div id="geoms" class="section level3">
<h3><span class="header-section-number">4.4.4</span> Geoms</h3>
<p>We can use <code>geoms</code> to specify how we would like data to be displayed on our graphs.
For example, our choice of geom could specify that we would like our data to
be displayed as a scatterplot, a barplot or a boxplot.</p>
<p>Let’s see how our graph would look as a scatterplot.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> cell1, <span class="dt">y =</span> cell2)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb27-2" title="2"><span class="st">  </span><span class="kw">geom_point</span>()</a></code></pre></div>
<p><img src="04-r-and-bioconductor_files/figure-html/unnamed-chunk-20-1.png" width="90%" style="display: block; margin: auto;" /></p>
<p>Now we can see that there doesn’t seem to be any correlation between
gene expression in <code>cell1</code> and <code>cell2</code>. Given we generated <code>counts</code> randomly,
this isn’t too surprising.</p>
<p>We can modify the command above to create a line plot using the <code>geom_line()</code>
function.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> cell1, <span class="dt">y =</span> cell2)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb28-2" title="2"><span class="st">  </span><span class="kw">geom_line</span>()</a></code></pre></div>
<p><img src="04-r-and-bioconductor_files/figure-html/unnamed-chunk-21-1.png" width="90%" style="display: block; margin: auto;" /></p>
<p>Hint: execute <code>?ggplot</code> and scroll down the help page. At the bottom is a
link to the <code>ggplot2</code> package index. Scroll through the index until you
find the <code>geom</code> options.</p>
</div>
<div id="plotting-data-from-more-than-2-cells" class="section level3">
<h3><span class="header-section-number">4.4.5</span> Plotting data from more than 2 cells</h3>
<p>So far we’ve been considering the gene counts from 2 of the cells in our
dataframe. But there are actually 10 cells in our dataframe and it would
be nice to compare all of them. What if we wanted to plot data from all
10 cells at the same time?</p>
<p>At the moment we can’t do this because we are treating each individual
cell as a variable and assigning that variable to either the x or the y
axis. We could create a 10 dimensional graph to plot data from all 10
cells on, but this is a) not possible to do with ggplot and b) not
very easy to interpret. What we could do instead is to tidy our data
so that we had one variable representing cell ID and another variable
representing gene counts, and plot those against each other.</p>
<p>In code, this would look like:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" title="1">counts &lt;-<span class="st"> </span><span class="kw">gather</span>(counts, <span class="kw">colnames</span>(counts)[<span class="dv">2</span><span class="op">:</span><span class="dv">11</span>], </a>
<a class="sourceLine" id="cb29-2" title="2">                 <span class="dt">key =</span> <span class="st">&#39;Cell_ID&#39;</span>, <span class="dt">value=</span><span class="st">&#39;Counts&#39;</span>)</a>
<a class="sourceLine" id="cb29-3" title="3"><span class="kw">head</span>(counts)</a></code></pre></div>
<pre><code>##   gene_ids Cell_ID Counts
## 1    gene1   cell1     11
## 2    gene2   cell1     12
## 3    gene3   cell1      9
## 4    gene4   cell1      8
## 5    gene5   cell1     11
## 6    gene6   cell1      4</code></pre>
<p>Essentially, the problem before was that our data was not tidy because
one variable (<code>Cell_ID</code>) was spread over multiple columns. Now that we’ve
fixed this problem, it is much easier for us to plot data from all
10 cells on one graph.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" title="1"><span class="kw">ggplot</span>(counts,<span class="kw">aes</span>(<span class="dt">x=</span>Cell_ID, <span class="dt">y=</span>Counts)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_boxplot</span>()</a></code></pre></div>
<p><img src="04-r-and-bioconductor_files/figure-html/unnamed-chunk-23-1.png" width="90%" style="display: block; margin: auto;" /></p>

</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-McCarthy2017-kb">
<p>McCarthy, Davis J., Kieran R. Campbell, Aaron T. L. Lun, and Quin F. Wills. 2017. “Scater: Pre-processing, Quality Control, Normalization and Visualization of Single-Cell RNA-Seq Data in R.” <em>Bioinformatics</em>, January, btw777. <a href="https://doi.org/10.1093/bioinformatics/btw777">https://doi.org/10.1093/bioinformatics/btw777</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="construction-of-expression-matrix.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="cleaning-the-expression-matrix.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["scRNA-seq-course.pdf"],
"toc": {
"collapse": "section"
},
"search": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
